{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Project Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import time\n",
    "from datetime import datetime\n",
    "from glob import glob\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import sfacts as sf\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "from lib.dissimilarity import load_dmat_as_pickle\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import lib.thisproject.data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"paper\")\n",
    "plt.rcParams[\"figure.dpi\"] = 300\n",
    "# NOTE: You can use the Chrome page inspector to make these large, presentable PNGs fit inside your notebook effectively, too.\n",
    "\n",
    "# FIXME: Trying to make figures with Helvetica or Arial as the font, but apparently these aren't in available locations...?\n",
    "plt.rcParams[\"font.sans-serif\"] = [\n",
    "    # \"Helvetica\",\n",
    "    \"DejaVu Sans\",\n",
    "    # \"Bitstream Vera Sans\",\n",
    "    # \"Computer Modern Sans Serif\",\n",
    "    # \"Lucida Grande\",\n",
    "    # \"Verdana\",\n",
    "    # \"Geneva\",\n",
    "    # \"Lucid\",\n",
    "    # \"Arial\",\n",
    "    # \"Avant Garde\",\n",
    "    # \"sans-serif\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = (\n",
    "    pd.read_table(\"meta/species_group.tsv\")[lambda x: x.species_group_id == \"hmp2\"]\n",
    "    .species_id.astype(str)\n",
    "    .unique()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_taxonomy_string(taxonomy_string):\n",
    "    values = taxonomy_string.split(\";\")\n",
    "    return pd.Series(values, index=[\"d__\", \"p__\", \"c__\", \"o__\", \"f__\", \"g__\", \"s__\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy_inpath = \"ref/uhgg_genomes_all_v2.tsv\"\n",
    "\n",
    "species_taxonomy = (\n",
    "    pd.read_table(species_taxonomy_inpath)[lambda x: x.Genome == x.Species_rep]\n",
    "    .assign(species_id=lambda x: \"1\" + x.MGnify_accession.str.split(\"-\").str[2])\n",
    "    .set_index(\"species_id\")\n",
    "    .Lineage.apply(parse_taxonomy_string)\n",
    ")\n",
    "species_taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_order = [\n",
    "    \"p__Euryarchaeota\",\n",
    "    \"p__Thermoplasmatota\",\n",
    "    \"p__Firmicutes\",\n",
    "    \"p__Firmicutes_A\",\n",
    "    \"p__Firmicutes_C\",\n",
    "    # \"p__Firmicutes_B\", # None in species_list1\n",
    "    # \"p__Firmicutes_G\", # B/G/I not sure how related to C or A\n",
    "    # \"p__Firmicutes_I\", #\n",
    "    # \"p__Cyanobacteria\", # None in species_list1\n",
    "    \"p__Actinobacteriota\",\n",
    "    \"p__Synergistota\",\n",
    "    \"p__Fusobacteriota\",\n",
    "    \"p__Campylobacterota\",\n",
    "    \"p__Proteobacteria\",\n",
    "    \"p__Desulfobacterota_A\",\n",
    "    \"p__Bacteroidota\",\n",
    "    \"p__Verrucomicrobiota\",\n",
    "    # \"dummy0\", # 18\n",
    "    # \"dummy1\", # 19\n",
    "    # \"dummy2\", # 20\n",
    "]\n",
    "\n",
    "phylum_palette = lib.plot.construct_ordered_palette(\n",
    "    phylum_order,\n",
    "    cm=\"rainbow\",\n",
    "    desaturate_levels=[1.0, 0.5],\n",
    ")\n",
    "\n",
    "for p__ in phylum_order:\n",
    "    print(p__, phylum_palette[p__])\n",
    "    plt.scatter([], [], color=phylum_palette[p__], label=p__)\n",
    "plt.legend(ncols=4)\n",
    "lib.plot.hide_axes_and_spines()\n",
    "\n",
    "# assert len(set(phylum_palette.values())) == len((phylum_palette.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_genome(x):\n",
    "    if (x.genome_type == \"Isolate\") & x.passes_filter:\n",
    "        return \"isolate\"\n",
    "    elif (x.genome_type == \"Isolate\") & ~x.passes_filter:\n",
    "        return \"isolate_fails_qc\"\n",
    "    elif (x.genome_type == \"MAG\") & x.passes_filter:\n",
    "        return \"mag\"\n",
    "    elif (x.genome_type == \"MAG\") & ~x.passes_filter:\n",
    "        return \"mag_fails_qc\"\n",
    "    elif (x.genome_type == \"SPGC\") & x.passes_filter:\n",
    "        return \"spgc\"\n",
    "    elif (x.genome_type == \"SPGC\") & x.passes_geno_positions:\n",
    "        return \"sfacts_only\"\n",
    "    elif (x.genome_type == \"SPGC\") & ~(x.passes_geno_positions):\n",
    "        return \"sfacts_fails_qc\"\n",
    "    else:\n",
    "        raise ValueError(\"Genome did not match classification criteria:\", x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats = []\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list\n",
    "# _species_list = [\"100003\"]\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.strain_meta_spgc_and_ref.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath).assign(species=species, inpath=inpath)\n",
    "    filt_stats.append(data)\n",
    "filt_stats = (\n",
    "    pd.concat(filt_stats)\n",
    "    .assign(\n",
    "        genome_class=lambda x: x.apply(classify_genome, axis=1),\n",
    "        species_strain=lambda x: x.species + \"_\" + x.genome_id,\n",
    "    )\n",
    "    .set_index(\"species_strain\")\n",
    ")\n",
    "\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define different subsets of the species:\n",
    "\n",
    "# All species:\n",
    "# species_list\n",
    "spgc_strain_list = filt_stats[lambda x: x.genome_type.isin([\"SPGC\"])].index.values\n",
    "\n",
    "# All species with enough positions\n",
    "species_list0 = filt_stats[lambda x: x.passes_geno_positions].species.unique()\n",
    "spgc_strain_list0 = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list0)\n",
    "].index.values\n",
    "\n",
    "# All species with sf strains\n",
    "species_list1 = filt_stats[\n",
    "    lambda x: x.passes_geno_positions & x.genome_type.isin([\"SPGC\"])\n",
    "].species.unique()\n",
    "spgc_strain_list1 = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list1)\n",
    "].index.values\n",
    "\n",
    "# All species with sf strains to talk about distributions (>=10)\n",
    "species_list1b = idxwhere(\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_geno_positions & x.genome_type.isin([\"SPGC\"])\n",
    "    ].species.value_counts()\n",
    "    >= 10\n",
    ")\n",
    "spgc_strain_list1b = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list1b)\n",
    "].index.values\n",
    "\n",
    "# All species with spgc strains\n",
    "species_list2 = filt_stats[\n",
    "    lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "].species.unique()\n",
    "spgc_strain_list2 = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list2)\n",
    "].index.values\n",
    "\n",
    "# All species with enough spgc strains for pangenome analysis (>=10)\n",
    "species_list3 = idxwhere(\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "    ].species.value_counts()\n",
    "    >= 10\n",
    ")\n",
    "spgc_strain_list3 = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list3)\n",
    "].index.values\n",
    "\n",
    "# Species with large numbers of strains (>=20)\n",
    "species_list4 = idxwhere(\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "    ].species.value_counts()\n",
    "    >= 20\n",
    ")\n",
    "spgc_strain_list4 = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list4)\n",
    "].index.values\n",
    "\n",
    "_species_list_map = {\n",
    "    \"All considered species\": (species_list, spgc_strain_list),\n",
    "    \"0: Species with enough genotyped positions\": (species_list0, spgc_strain_list0),\n",
    "    \"1: With sfacts strains\": (species_list1, spgc_strain_list1),\n",
    "    \"1b: With (>=10) sfacts strains\": (species_list1b, spgc_strain_list1b),\n",
    "    \"2: With SPGC inferences\": (species_list2, spgc_strain_list2),\n",
    "    \"3: With >=10 inferences\": (species_list3, spgc_strain_list3),\n",
    "    \"4: With >=20 inferences\": (species_list4, spgc_strain_list4),\n",
    "}\n",
    "for _species_list_name, (_species_list, _strain_list) in _species_list_map.items():\n",
    "    print(_species_list_name, len(_species_list), len(_strain_list))\n",
    "    print(species_taxonomy.loc[_species_list].p__.value_counts())\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "species = '103166'\n",
    "\n",
    "cog_category_inpath = f'data/species/sp-{species}/midasdb_v15.emapper.gene75_x_cog_category.tsv'\n",
    "eggnog_inpath = f'data/species/sp-{species}/midasdb_v15.emapper.gene75_x_eggnog.tsv'\n",
    "ko_inpath = f'data/species/sp-{species}/midasdb_v15.emapper.gene75_x_ko.tsv'\n",
    "amr_inpath = f'data/species/sp-{species}/midasdb_v15.gene75_x_amr.tsv'\n",
    "plasmid_inpath = f'data/species/sp-{species}/midasdb_v15.gene75_x_genomad_plasmid.tsv'\n",
    "phage_inpath = f'data/species/sp-{species}/midasdb_v15.gene75_x_genomad_virus.tsv'\n",
    "emapper_inpath = f'ref/midasdb_uhgg_v15/pangenomes/{species}/eggnog.tsv'\n",
    "clust_inpath = f'data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.uhgg-strain_gene.gene_clust-t10.tsv'\n",
    "gene_family_inpath = f'ref/midasdb_uhgg_v15/pangenomes/{species}/gene_info.txt'\n",
    "morans_i_inpath = f'data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.uhgg-strain_gene.morans_i.tsv'\n",
    "prevalence_inpath = f'data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.uhgg-strain_gene.prevalence.tsv'\n",
    "kegg_module_inpath = f'data/species/sp-{species}/midasdb_v15.emapper.gene75_x_kegg_module.tsv'\n",
    "\n",
    "pdist_inpath = f'data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.spgc_ss-all.geno_uhgg-v15_pdist-mask10-pseudo10.pkl'\n",
    "strain_gene_inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.uhgg-strain_gene.tsv\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_family = pd.read_table(gene_family_inpath, index_col=\"gene_id\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "kegg_module = pd.read_table(kegg_module_inpath)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "emapper_all = pd.read_table(emapper_inpath, index_col='#query').rename_axis('centroid_99')\n",
    "description = emapper_all.reindex(gene_family.centroid_75.unique()).dropna(subset=['seed_ortholog']).rename_axis('centroid_75')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "phage = pd.read_table(phage_inpath)#.groupby('centroid_75')\n",
    "plasmid = pd.read_table(plasmid_inpath)\n",
    "amr = pd.read_table(amr_inpath)\n",
    "\n",
    "clust = pd.read_table(clust_inpath, names=['centroid_75', 'clust'], index_col='centroid_75').clust\n",
    "clust_size = clust.value_counts()\n",
    "# clust = clust[lambda x: (x >= 0) & x.isin(idxwhere(clust_size > 1))]\n",
    "\n",
    "morans_i = pd.read_table(morans_i_inpath, names=['centroid_75', 'morans_i'], index_col='centroid_75').morans_i.dropna()\n",
    "prevalence = pd.read_table(prevalence_inpath, names=['centroid_75', 'prevalence'], index_col='centroid_75').prevalence"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "eggnog = pd.read_table(eggnog_inpath)\n",
    "cog_category = pd.read_table(cog_category_inpath)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "from scipy.cluster.hierarchy import linkage\n",
    "from scipy.spatial.distance import squareform\n",
    "\n",
    "strain_gene = pd.read_table(strain_gene_inpath, index_col='gene_id')\n",
    "geno_dmat = lib.dissimilarity.load_dmat_as_pickle(pdist_inpath).loc[strain_gene.columns, strain_gene.columns]\n",
    "geno_linkage = linkage(squareform(geno_dmat), optimal_ordering=True)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "plt.hist(morans_i, bins=100)\n",
    "None"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d = description[['Preferred_name', 'Description']].reindex(gene_family.centroid_75.unique()).assign(morans_i=morans_i).assign(\n",
    "    cog_category=cog_category.groupby('centroid_75').cog_category.apply(''.join),\n",
    "    eggnog=eggnog.groupby('centroid_75').eggnog.apply(';'.join),\n",
    "    kegg_module=kegg_module.groupby('centroid_75').kegg_module.apply(';'.join),\n",
    "    clust=clust,\n",
    "    csize=lambda x: x.clust.map(clust_size),\n",
    "    plasmid=plasmid.groupby('centroid_75').annotation_accessions.apply(';'.join),\n",
    "    phage=phage.groupby('centroid_75').annotation_accessions.apply(';'.join),\n",
    "    amr=amr.groupby('centroid_75').accession_no.apply(';'.join),\n",
    "    prevalence=prevalence,\n",
    ")\n",
    "\n",
    "# d[lambda x: ~x.clust.isna() & (prevalence > 0.1) & (prevalence < 0.9)].sort_values('csize', ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# clust_inpath = f'data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.uhgg-strain_gene.gene_clust-t10.tsv'\n",
    "# kegg_module_inpath = f'data/species/sp-{species}/midasdb_v15.emapper.gene75_x_kegg_module.tsv'\n",
    "# clust = pd.read_table(clust_inpath, names=['centroid_75', 'clust'], index_col='centroid_75').clust\n",
    "# clust_size = clust.value_counts()\n",
    "# kegg_module = pd.read_table(kegg_module_inpath)\n",
    "\n",
    "\n",
    "x = kegg_module.join(clust, on='centroid_75')[['clust', 'kegg_module']][lambda x: x.clust >= 0]\n",
    "non_singleton_clusters = idxwhere(clust_size > 1)\n",
    "\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "# ax2 = plt.twinx(ax1)\n",
    "# ax2.invert_yaxis()\n",
    "\n",
    "xmax = 15\n",
    "bins = np.arange(xmax) - 0.5\n",
    "\n",
    "# ax.hist(x[lambda y: y.clust.isin(non_singleton_clusters)].value_counts(), bins=np.arange(xmax) - 0.5, alpha=0.5, align='mid', color='tab:blue')\n",
    "obs_counts = x[lambda y: y.clust.isin(non_singleton_clusters)].value_counts()\n",
    "ax.hist(obs_counts, bins=bins, alpha=1.0, histtype='step', lw=2, align='mid', color='tab:orange', density=False, label='__nolegend__')\n",
    "\n",
    "\n",
    "np.random.seed(0)\n",
    "n_reps = 100\n",
    "all_perm_counts = []\n",
    "for i in range(n_reps):\n",
    "    x_perm = x.assign(clust=lambda x: x.clust.sample(frac=1).values)\n",
    "    x_counts = x_perm[lambda y: y.clust.isin(non_singleton_clusters)].value_counts()\n",
    "    all_perm_counts.append(x_counts)\n",
    "    ax.hist(x_counts, bins=bins, alpha=0.8 / n_reps, align='mid', color='k', density=False, label='__nolegend__')\n",
    "\n",
    "ax.scatter([], [], edgecolor='tab:orange', facecolor='none', lw=2, label='observed', marker='s', s=200)\n",
    "ax.scatter([], [], c='black', alpha=0.2, label='null', marker='s', s=200, lw=0)\n",
    "ax.legend(loc='upper right')\n",
    "\n",
    "ax.set_xticks([1, 3, 5, 7, 9, 11])\n",
    "ax.set_xlabel('Genes per Module')\n",
    "ax.set_ylabel('Cluster-Modules (count)')\n",
    "\n",
    "ax.set_yscale('symlog', linthresh=1, linscale=0.1)\n",
    "\n",
    "# # ax1.set_yscale('symlog')\n",
    "# # ax2.set_yscale('symlog')\n",
    "# ax.set_ylim(0, 1000)\n",
    "# ax.set_yticks(np.array([1e-3, 1e-2, 1e-1, 1]))\n",
    "\n",
    "x[lambda y: y.clust.isin(non_singleton_clusters)].value_counts().to_frame('tally').reset_index().join(clust_size, on='clust').head(20)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Simple summary of permutation test results for modules\n",
    "\n",
    "for min_num_counts in [2, 3, 4, 5, 6, 7, 8]:\n",
    "    print(\n",
    "        # Threshold num genes\n",
    "        min_num_counts,\n",
    "        # Observed num clusters\n",
    "        (obs_counts >= min_num_counts).sum(),\n",
    "        # Max result under permutation\n",
    "        np.array([(perm_counts >= min_num_counts).sum() for perm_counts in all_perm_counts]).max(),\n",
    "    )"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# clust_inpath = f'data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.uhgg-strain_gene.gene_clust-t10.tsv'\n",
    "# kegg_module_inpath = f'data/species/sp-{species}/midasdb_v15.emapper.gene75_x_kegg_module.tsv'\n",
    "# cog_category_inpath = f'data/species/sp-{species}/midasdb_v15.emapper.gene75_x_cog_category.tsv'\n",
    "# clust = pd.read_table(clust_inpath, names=['centroid_75', 'clust'], index_col='centroid_75').clust\n",
    "# clust_size = clust.value_counts()\n",
    "# cog_category = pd.read_table(cog_category_inpath)\n",
    "\n",
    "\n",
    "x = cog_category[lambda x: x.cog_category != 'no_category'].join(clust, on='centroid_75')[['clust', 'cog_category']][lambda x: x.clust >= 0]\n",
    "non_singleton_clusters = idxwhere(clust_size > 1)\n",
    "\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "# ax2 = plt.twinx(ax1)\n",
    "# ax2.invert_yaxis()\n",
    "\n",
    "xmax = 15\n",
    "bins = np.arange(xmax) - 0.5\n",
    "\n",
    "# ax.hist(x[lambda y: y.clust.isin(non_singleton_clusters)].value_counts(), bins=np.arange(xmax) - 0.5, alpha=0.5, align='mid', color='tab:blue')\n",
    "obs_counts = x[lambda y: y.clust.isin(non_singleton_clusters)].value_counts()\n",
    "ax.hist(obs_counts, bins=bins, alpha=1.0, histtype='step', lw=2, align='mid', color='tab:orange', density=False, label='__nolegend__')\n",
    "\n",
    "\n",
    "np.random.seed(0)\n",
    "n_reps = 100\n",
    "all_perm_counts = []\n",
    "for i in range(n_reps):\n",
    "    x_perm = x.assign(clust=lambda x: x.clust.sample(frac=1).values)\n",
    "    x_counts = x_perm[lambda y: y.clust.isin(non_singleton_clusters)].value_counts()\n",
    "    all_perm_counts.append(x_counts)\n",
    "    ax.hist(x_counts, bins=bins, alpha=0.8 / n_reps, align='mid', color='k', density=False, label='__nolegend__')\n",
    "\n",
    "ax.scatter([], [], edgecolor='tab:orange', facecolor='none', lw=2, label='observed', marker='s', s=200)\n",
    "ax.scatter([], [], c='black', alpha=0.2, label='null', marker='s', s=200, lw=0)\n",
    "ax.legend(loc='upper right')\n",
    "\n",
    "ax.set_xticks([1, 3, 5, 7, 9, 11])\n",
    "ax.set_xlabel('Genes per Module')\n",
    "ax.set_ylabel('Cluster-Categories (count)')\n",
    "\n",
    "ax.set_yscale('symlog', linthresh=1, linscale=0.1)\n",
    "\n",
    "# # ax1.set_yscale('symlog')\n",
    "# # ax2.set_yscale('symlog')\n",
    "# ax.set_ylim(0, 1000)\n",
    "# ax.set_yticks(np.array([1e-3, 1e-2, 1e-1, 1]))\n",
    "\n",
    "x[lambda y: y.clust.isin(non_singleton_clusters)].value_counts().to_frame('tally').reset_index().join(clust_size, on='clust').head(20)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Simple summary of permutation test results for modules\n",
    "\n",
    "for min_num_counts in [2, 3, 4, 5, 6, 7, 8]:\n",
    "    print(\n",
    "        # Threshold num genes\n",
    "        min_num_counts,\n",
    "        # Observed num clusters\n",
    "        (obs_counts >= min_num_counts).sum(),\n",
    "        # Max result under permutation\n",
    "        np.array([(perm_counts >= min_num_counts).sum() for perm_counts in all_perm_counts]).max(),\n",
    "    )"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d[lambda x: x.clust == 2070].head(50)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sns.clustermap(strain_gene[clust == 2070], col_linkage=geno_linkage)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Higher rate of same-categories and same-module (plasmids?) matching in clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather clusters for all species (among a particular species list, that is)\n",
    "# Gather annotations for all genes in these clusters\n",
    "# Put everything together and do the same analysis as above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clust = []\n",
    "kegg_module = []\n",
    "cog_category = []\n",
    "plasmid = []\n",
    "phage = []\n",
    "\n",
    "for species in tqdm(species_list4):\n",
    "    clust_inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.uhgg-strain_gene.gene_clust-t10.tsv\"\n",
    "    kegg_module_inpath = (\n",
    "        f\"data/species/sp-{species}/midasdb_v15.emapper.gene75_x_kegg_module.tsv\"\n",
    "    )\n",
    "    cog_category_inpath = (\n",
    "        f\"data/species/sp-{species}/midasdb_v15.emapper.gene75_x_cog_category.tsv\"\n",
    "    )\n",
    "    plasmid_inpath = (\n",
    "        f\"data/species/sp-{species}/midasdb_v15.gene75_x_genomad_plasmid.tsv\"\n",
    "    )\n",
    "    phage_inpath = f\"data/species/sp-{species}/midasdb_v15.gene75_x_genomad_virus.tsv\"\n",
    "    clust.append(\n",
    "        pd.read_table(clust_inpath, names=[\"centroid_75\", \"clust\"]).assign(\n",
    "            species=species\n",
    "        )\n",
    "    )\n",
    "    kegg_module.append(pd.read_table(kegg_module_inpath))\n",
    "    cog_category.append(pd.read_table(cog_category_inpath))\n",
    "    plasmid.append(pd.read_table(plasmid_inpath))\n",
    "    phage.append(pd.read_table(phage_inpath))\n",
    "\n",
    "clust = pd.concat(clust).assign(\n",
    "    clust_label=lambda x: x.clust.astype(str) + \"_\" + x.species\n",
    ")\n",
    "kegg_module = pd.concat(kegg_module)\n",
    "cog_category = pd.concat(cog_category)\n",
    "plasmid = pd.concat(plasmid)\n",
    "phage = pd.concat(phage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many clusters (with >=2 genes)?\n",
    "clust[lambda x: (x.clust >= 0)].clust_label.value_counts()[lambda x: x > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Median number of clusters per species.\n",
    "clust[lambda x: (x.clust >= 0)].groupby(\"species\").clust_label.value_counts()[\n",
    "    lambda x: x > 1\n",
    "].reset_index().species.value_counts().median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.array([2**i for i in range(1, 11)])\n",
    "\n",
    "d = (\n",
    "    clust[lambda x: x.clust >= 0]\n",
    "    .groupby(\"species\")\n",
    "    .apply(\n",
    "        lambda d: pd.Series(\n",
    "            np.histogram(d.clust_label.value_counts(), bins=bins)[0], index=bins[:-1]\n",
    "        )\n",
    "    )\n",
    ")\n",
    "# Order by total number of genes in clusters with size \u22652\n",
    "species_order = (\n",
    "    (\n",
    "        clust[lambda x: x.clust >= 0][[\"species\", \"clust\"]].value_counts()[\n",
    "            lambda x: x > 1\n",
    "        ]\n",
    "    )\n",
    "    .groupby(\"species\")\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)\n",
    "    .index\n",
    ")\n",
    "\n",
    "row_colors = (\n",
    "    d.index.to_series().map(species_taxonomy.p__).map(phylum_palette)#.rename(\"\")\n",
    ")\n",
    "\n",
    "norm = mpl.colors.SymLogNorm(1, vmin=0, vmax=2_000)\n",
    "cmap = sns.color_palette(\"rocket\", as_cmap=True)\n",
    "\n",
    "cg = sns.clustermap(\n",
    "    d.loc[species_order].drop(columns=[1], errors=\"ignore\"),\n",
    "    norm=norm,\n",
    "    cmap=cmap,\n",
    "    xticklabels=1,\n",
    "    yticklabels=0,\n",
    "    col_cluster=False,\n",
    "    row_cluster=False,\n",
    "    row_colors=row_colors,\n",
    "    figsize=(4.5, 3.5),\n",
    "    cbar_pos=None,\n",
    ")\n",
    "cg.ax_heatmap.set_ylabel(\"\")\n",
    "cg.figure.colorbar(mpl.cm.ScalarMappable(norm=norm, cmap=cmap), ax=cg.ax_heatmap)\n",
    "cg.ax_heatmap.set_xticks(\n",
    "    ticks=np.array(cg.ax_heatmap.get_xticks()) + 0.45, labels=d.columns, rotation=0\n",
    ")\n",
    "\n",
    "# fig, ax = plt.subplots(figsize=(0.15, 1.5), facecolor='none')\n",
    "# fig.colorbar(\n",
    "# # lib.plot.hide_axes_and_spines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.array([2**i for i in range(1, 11)])\n",
    "\n",
    "d = (\n",
    "    clust[lambda x: x.clust >= 0]\n",
    "    .groupby(\"species\")\n",
    "    .apply(\n",
    "        lambda d: pd.Series(\n",
    "            np.histogram(d.clust_label.value_counts(), bins=bins)[0], index=bins[:-1]\n",
    "        )\n",
    "    )\n",
    ")\n",
    "# Order by total number of genes in clusters with size \u22652\n",
    "species_order = (\n",
    "    (\n",
    "        clust[lambda x: x.clust >= 0][[\"species\", \"clust\"]].value_counts()[\n",
    "            lambda x: x > 1\n",
    "        ]\n",
    "    )\n",
    "    .groupby(\"species\")\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)\n",
    "    .index\n",
    ")\n",
    "\n",
    "row_colors = (\n",
    "    d.index.to_series().map(species_taxonomy.p__).map(phylum_palette).rename(\"\")\n",
    ")\n",
    "\n",
    "norm = mpl.colors.SymLogNorm(1, vmin=0, vmax=2_000)\n",
    "cmap = sns.color_palette(\"rocket\", as_cmap=True)\n",
    "\n",
    "cg = sns.clustermap(\n",
    "    d.loc[species_order].drop(columns=[1], errors=\"ignore\").T,\n",
    "    norm=norm,\n",
    "    cmap=cmap,\n",
    "    xticklabels=0,\n",
    "    yticklabels=1,\n",
    "    col_cluster=False,\n",
    "    row_cluster=False,\n",
    "    col_colors=row_colors,\n",
    "    figsize=(6, 3),\n",
    "    cbar_pos=None,\n",
    ")\n",
    "cg.ax_heatmap.set_yticks(\n",
    "    ticks=np.array(cg.ax_heatmap.get_yticks()) - 0.45, labels=d.columns, rotation=0\n",
    ")\n",
    "# lib.plot.rotate_yticklabels(ax=cg.ax_heatmap, rotation=0, va='center')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(0.15, 1.5), facecolor=\"none\")\n",
    "fig.colorbar(mpl.cm.ScalarMappable(norm=norm, cmap=cmap), cax=ax)\n",
    "# lib.plot.hide_axes_and_spines()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(1.5, 0.15), facecolor=\"none\")\n",
    "fig.colorbar(\n",
    "    mpl.cm.ScalarMappable(norm=norm, cmap=cmap), cax=ax, orientation=\"horizontal\"\n",
    ")\n",
    "# lib.plot.hide_axes_and_spines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(2, 5), facecolor=\"none\")\n",
    "plt.scatter(\n",
    "    [],\n",
    "    [],\n",
    "    c=[],\n",
    "    norm=mpl.colors.SymLogNorm(1, vmin=0, vmax=2_000),\n",
    "    cmap=sns.color_palette(\"rocket\", as_cmap=True),\n",
    ")\n",
    "plt.colorbar()\n",
    "lib.plot.hide_axes_and_spines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_singleton_clusters = idxwhere(\n",
    "    (clust[lambda x: x.clust >= 0].clust_label.value_counts() > 1)\n",
    ")\n",
    "len(non_singleton_clusters)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### KEGG Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_annot = kegg_module.rename(columns={\"kegg_module\": \"annot\"})\n",
    "\n",
    "all_counted_genes_clust_info = clust[lambda x: x.clust > 0]\n",
    "obs_clust_annot_count = (\n",
    "    pd.merge(all_counted_genes_clust_info, _annot, on=\"centroid_75\", how=\"left\")[\n",
    "        [\"clust_label\", \"annot\"]\n",
    "    ]\n",
    "    .value_counts()\n",
    "    .groupby(\"clust_label\")\n",
    "    .max()\n",
    ")\n",
    "\n",
    "np.random.seed(0)\n",
    "n_perm = 100\n",
    "perm_clust_annot_count_list = []\n",
    "for i in tqdm(range(n_perm)):\n",
    "    perm_clust_annot_count = (\n",
    "        pd.merge(\n",
    "            all_counted_genes_clust_info.groupby(\"species\").apply(\n",
    "                lambda d: d.assign(\n",
    "                    clust_label=lambda x: x.clust_label.sample(frac=1).values\n",
    "                )\n",
    "            ),\n",
    "            _annot,\n",
    "            on=\"centroid_75\",\n",
    "            how=\"left\",\n",
    "        )[[\"clust_label\", \"annot\"]]\n",
    "        .value_counts()\n",
    "        .groupby(\"clust_label\")\n",
    "        .max()\n",
    "    )\n",
    "    perm_clust_annot_count_list.append(perm_clust_annot_count)\n",
    "\n",
    "\n",
    "bins = np.arange(obs_clust_annot_count.max()) + 1\n",
    "\n",
    "perm_histogram2d = pd.DataFrame(\n",
    "    np.stack(\n",
    "        [\n",
    "            np.histogram(perm_clust_annot_count, bins=bins)[0]\n",
    "            for perm_clust_annot_count in perm_clust_annot_count_list\n",
    "        ]\n",
    "    ),\n",
    "    columns=bins[:-1],\n",
    ")\n",
    "obs_histogram = pd.Series(\n",
    "    np.histogram(obs_clust_annot_count, bins=bins)[0], index=bins[:-1]\n",
    ")\n",
    "sns.clustermap(\n",
    "    perm_histogram2d,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=1),\n",
    "    row_cluster=False,\n",
    "    col_cluster=False,\n",
    ")\n",
    "\n",
    "kegg_module_bins = bins\n",
    "kegg_module_perm_clust_annot_count = perm_clust_annot_count\n",
    "kegg_module_obs_clust_annot_count = obs_clust_annot_count\n",
    "kegg_module_perm_histogram2d = perm_histogram2d\n",
    "kegg_module_obs_histogram = obs_histogram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(4, 2))\n",
    "\n",
    "_perm = kegg_module_perm_histogram2d\n",
    "_obs = kegg_module_obs_clust_annot_count\n",
    "_bins = kegg_module_bins\n",
    "\n",
    "# # Plot the permutation histograms\n",
    "# for perm_clust_annot_count in perm_clust_annot_count_list:\n",
    "#     ax.hist(perm_clust_annot_count, bins=bins, align='left', color='k', alpha=0.5 / n_perm)\n",
    "# # Plot the expected values.\n",
    "# ax.stairs(perm_histogram2d.mean(0), bins - 0.5, edgecolor='k', facecolor='none', lw=1)\n",
    "\n",
    "# Plot the mean\n",
    "ax.stairs(\n",
    "    _perm.mean(0),\n",
    "    _bins - 0.5,\n",
    "    edgecolor=\"none\",\n",
    "    facecolor=\"k\",\n",
    "    lw=1,\n",
    "    fill=True,\n",
    "    alpha=0.2,\n",
    ")\n",
    "ax.stairs(\n",
    "    _perm.mean(0),\n",
    "    _bins - 0.5,\n",
    "    edgecolor=\"k\",\n",
    "    facecolor=\"none\",\n",
    "    lw=1,\n",
    "    fill=True,\n",
    "    alpha=1.0,\n",
    ")\n",
    "\n",
    "# ax.hist(perm_histogram2d.mean(0), bins=bins, align='left', color='k', alpha=0.5)\n",
    "# ax.hist(perm_histogram2d.mean(0), bins=bins, align='left', color='k', histtype='step')\n",
    "\n",
    "ax.hist(_obs, bins=_bins, align=\"left\", histtype=\"step\", color=\"tab:orange\")\n",
    "# ax.hist(obs_clust_annot_count, bins=bins, align='left', histtype=\"stepfilled\", color='tab:orange', alpha=0.2)\n",
    "ax.set_yscale(\"symlog\", linthresh=1, linscale=0.1)\n",
    "\n",
    "ax.scatter(\n",
    "    [],\n",
    "    [],\n",
    "    edgecolor=\"tab:orange\",\n",
    "    facecolor=\"none\",\n",
    "    lw=1,\n",
    "    label=\"observed\",\n",
    "    marker=\"s\",\n",
    "    s=40,\n",
    ")\n",
    "ax.scatter(\n",
    "    [], [], edgecolor=\"k\", facecolor=\"silver\", label=\"null\", marker=\"s\", s=40, lw=1\n",
    ")\n",
    "ax.legend(loc=\"upper right\")\n",
    "\n",
    "ax.set_xticks(\n",
    "    np.concatenate([np.arange(1, 8, step=1), np.arange(10, _obs.max() + 1, step=5)])\n",
    ")\n",
    "ax.set_xlabel(\"Max Annotations\")\n",
    "ax.set_ylabel(\"Clusters (count)\")\n",
    "ax.set_ylim(0, 1e5)\n",
    "ax.set_yticks(np.logspace(0, 5, num=6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_perm = kegg_module_perm_histogram2d\n",
    "_obs = kegg_module_obs_histogram\n",
    "\n",
    "obs_num_clust_ge3 = _obs.iloc[::-1].cumsum().loc[3]\n",
    "null_num_clust_ge3 = _perm.iloc[:, ::-1].cumsum(1)[3]\n",
    "\n",
    "plt.hist(null_num_clust_ge3)\n",
    "plt.axvline(obs_num_clust_ge3, color=\"k\")\n",
    "\n",
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        obs=_obs.iloc[::-1].cumsum().iloc[::-1],\n",
    "        perm=_perm.iloc[:, ::-1].cumsum(1).iloc[:, ::-1].mean(),\n",
    "    )\n",
    ").assign(ratio=lambda x: x.obs / x.perm)\n",
    "d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plasmid / Phage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_annot = (\n",
    "    pd.DataFrame(\n",
    "        dict(\n",
    "            phage=phage.assign(phage=True)[[\"centroid_75\", \"phage\"]]\n",
    "            .drop_duplicates()\n",
    "            .set_index(\"centroid_75\")\n",
    "            .phage,\n",
    "            plasmid=plasmid.assign(plasmid=True)[[\"centroid_75\", \"plasmid\"]]\n",
    "            .drop_duplicates()\n",
    "            .set_index(\"centroid_75\")\n",
    "            .plasmid,\n",
    "        )\n",
    "    )\n",
    "    .stack()\n",
    "    .sort_values()\n",
    "    .rename_axis([\"centroid_75\", \"annot\"])\n",
    "    .reset_index()\n",
    "    .drop(columns=[0])\n",
    ")\n",
    "\n",
    "all_counted_genes_clust_info = clust[lambda x: x.clust > 0]\n",
    "obs_clust_annot_count = (\n",
    "    pd.merge(all_counted_genes_clust_info, _annot, on=\"centroid_75\", how=\"left\")[\n",
    "        [\"clust_label\", \"annot\"]\n",
    "    ]\n",
    "    .value_counts()\n",
    "    .groupby(\"clust_label\")\n",
    "    .max()\n",
    ")\n",
    "\n",
    "np.random.seed(0)\n",
    "n_perm = 100\n",
    "perm_clust_annot_count_list = []\n",
    "for i in tqdm(range(n_perm)):\n",
    "    perm_clust_annot_count = (\n",
    "        pd.merge(\n",
    "            all_counted_genes_clust_info.groupby(\"species\").apply(\n",
    "                lambda d: d.assign(\n",
    "                    clust_label=lambda x: x.clust_label.sample(frac=1).values\n",
    "                )\n",
    "            ),\n",
    "            _annot,\n",
    "            on=\"centroid_75\",\n",
    "            how=\"left\",\n",
    "        )[[\"clust_label\", \"annot\"]]\n",
    "        .value_counts()\n",
    "        .groupby(\"clust_label\")\n",
    "        .max()\n",
    "    )\n",
    "    perm_clust_annot_count_list.append(perm_clust_annot_count)\n",
    "\n",
    "\n",
    "bins = np.arange(obs_clust_annot_count.max()) + 1\n",
    "\n",
    "perm_histogram2d = pd.DataFrame(\n",
    "    np.stack(\n",
    "        [\n",
    "            np.histogram(perm_clust_annot_count, bins=bins)[0]\n",
    "            for perm_clust_annot_count in perm_clust_annot_count_list\n",
    "        ]\n",
    "    ),\n",
    "    columns=bins[:-1],\n",
    ")\n",
    "obs_histogram = pd.Series(\n",
    "    np.histogram(obs_clust_annot_count, bins=bins)[0], index=bins[:-1]\n",
    ")\n",
    "sns.clustermap(\n",
    "    perm_histogram2d,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=1),\n",
    "    row_cluster=False,\n",
    "    col_cluster=False,\n",
    ")\n",
    "\n",
    "plasmid_phage_bins = bins\n",
    "plasmid_phage_perm_clust_annot_count = perm_clust_annot_count\n",
    "plasmid_phage_obs_clust_annot_count = obs_clust_annot_count\n",
    "plasmid_phage_perm_histogram2d = perm_histogram2d\n",
    "plasmid_phage_obs_histogram = obs_histogram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(4, 2))\n",
    "\n",
    "_perm = plasmid_phage_perm_histogram2d\n",
    "_obs = plasmid_phage_obs_clust_annot_count\n",
    "_bins = plasmid_phage_bins\n",
    "\n",
    "# # Plot the permutation histograms\n",
    "# for perm_clust_annot_count in perm_clust_annot_count_list:\n",
    "#     ax.hist(perm_clust_annot_count, bins=bins, align='left', color='k', alpha=0.5 / n_perm)\n",
    "# # Plot the expected values.\n",
    "# ax.stairs(perm_histogram2d.mean(0), bins - 0.5, edgecolor='k', facecolor='none', lw=1)\n",
    "\n",
    "# Plot the mean\n",
    "ax.stairs(\n",
    "    _perm.mean(0),\n",
    "    _bins - 0.5,\n",
    "    edgecolor=\"none\",\n",
    "    facecolor=\"k\",\n",
    "    lw=1,\n",
    "    fill=True,\n",
    "    alpha=0.2,\n",
    ")\n",
    "ax.stairs(\n",
    "    _perm.mean(0),\n",
    "    _bins - 0.5,\n",
    "    edgecolor=\"k\",\n",
    "    facecolor=\"none\",\n",
    "    lw=1,\n",
    "    fill=True,\n",
    "    alpha=1.0,\n",
    ")\n",
    "\n",
    "# ax.hist(perm_histogram2d.mean(0), bins=bins, align='left', color='k', alpha=0.5)\n",
    "# ax.hist(perm_histogram2d.mean(0), bins=bins, align='left', color='k', histtype='step')\n",
    "\n",
    "ax.hist(_obs, bins=_bins, align=\"left\", histtype=\"step\", color=\"tab:orange\")\n",
    "# ax.hist(obs_clust_annot_count, bins=bins, align='left', histtype=\"stepfilled\", color='tab:orange', alpha=0.2)\n",
    "ax.set_yscale(\"symlog\", linthresh=1, linscale=0.1)\n",
    "\n",
    "ax.scatter(\n",
    "    [],\n",
    "    [],\n",
    "    edgecolor=\"tab:orange\",\n",
    "    facecolor=\"none\",\n",
    "    lw=1,\n",
    "    label=\"observed\",\n",
    "    marker=\"s\",\n",
    "    s=40,\n",
    ")\n",
    "ax.scatter(\n",
    "    [], [], edgecolor=\"k\", facecolor=\"silver\", label=\"null\", marker=\"s\", s=40, lw=1\n",
    ")\n",
    "ax.legend(loc=\"upper right\")\n",
    "\n",
    "ax.set_xticks(\n",
    "    np.concatenate([np.arange(1, 8, step=2), np.arange(10, _obs.max() + 1, step=5)])\n",
    ")\n",
    "ax.set_xlabel(\"Max Annotations\")\n",
    "ax.set_ylabel(\"Clusters (count)\")\n",
    "ax.set_ylim(0, 1e5)\n",
    "ax.set_yticks(np.logspace(0, 5, num=6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_perm = plasmid_phage_perm_histogram2d\n",
    "_obs = plasmid_phage_obs_histogram\n",
    "\n",
    "obs_num_clust_ge3 = _obs.iloc[::-1].cumsum().loc[3]\n",
    "null_num_clust_ge3 = _perm.iloc[:, ::-1].cumsum(1)[3]\n",
    "\n",
    "plt.hist(null_num_clust_ge3)\n",
    "plt.axvline(obs_num_clust_ge3, color=\"k\")\n",
    "\n",
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        obs=_obs.iloc[::-1].cumsum().iloc[::-1],\n",
    "        perm=_perm.iloc[:, ::-1].cumsum(1).iloc[:, ::-1].mean(),\n",
    "    )\n",
    ").assign(ratio=lambda x: x.obs / x.perm)\n",
    "d"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "bins = np.arange(obs_clust_annot_count.max()) + 1\n",
    "\n",
    "perm_histogram2d = pd.DataFrame(np.stack([np.histogram(perm_clust_annot_count, bins=bins)[0] for perm_clust_annot_count in perm_clust_annot_count_list]), columns=bins[:-1])\n",
    "obs_histogram = pd.Series(np.histogram(obs_clust_annot_count, bins=bins)[0], index=bins[:-1])\n",
    "\n",
    "sns.clustermap(perm_histogram2d, norm=mpl.colors.SymLogNorm(linthresh=1), row_cluster=False, col_cluster=False, )"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "fig, ax = plt.subplots(figsize=(6, 3))\n",
    "\n",
    "for perm_clust_annot_count in perm_clust_annot_count_list:\n",
    "    ax.hist(perm_clust_annot_count, bins=bins, align='left', color='k', alpha=0.5 / n_perm)\n",
    "\n",
    "ax.stairs(perm_histogram2d.mean(0), bins - 0.5, edgecolor='k', facecolor='none', lw=1)\n",
    "\n",
    "ax.hist(obs_clust_annot_count, bins=bins, align='left', histtype=\"step\", color='tab:orange', )\n",
    "ax.set_yscale('symlog', linthresh=1, linscale=0.1)\n",
    "\n",
    "ax.scatter([], [], edgecolor='tab:orange', facecolor='none', lw=2, label='observed', marker='s', s=200)\n",
    "ax.scatter([], [], edgecolor='k', facecolor='silver', label='null', marker='s', s=200, lw=1)\n",
    "ax.legend(loc='upper right')\n",
    "\n",
    "ax.set_xticks(np.concatenate([np.arange(1, 8, step=2), np.arange(10, obs_clust_annot_count.max() + 1, step=5)]))\n",
    "ax.set_xlabel('Phage or Plasmid Annotations')\n",
    "ax.set_ylabel('Clusters (count)')\n",
    "ax.set_ylim(0, 1e5)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d = pd.DataFrame(dict(obs=obs_histogram.iloc[::-1].cumsum().iloc[::-1], perm=perm_histogram2d.iloc[:,::-1].cumsum(1).iloc[:,::-1].mean())).assign(ratio=lambda x: x.obs / x.perm)\n",
    "d"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "obs_num_clust_ge3 = obs_histogram.iloc[::-1].cumsum()[3]\n",
    "null_num_clust_ge3 = perm_histogram2d.iloc[:,::-1].cumsum(1)[3]\n",
    "\n",
    "plt.hist(null_num_clust_ge3)\n",
    "plt.axvline(obs_num_clust_ge3, color='k')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "toolz4",
   "language": "python",
   "name": "toolz4"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}