{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Project Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import time\n",
    "from datetime import datetime\n",
    "from glob import glob\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"paper\")\n",
    "plt.rcParams[\"figure.dpi\"] = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpl.scale.register_scale(mpl.scale.FuncScale)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load and Check Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_list = idxwhere(\n",
    "    pd.read_table(\"meta/genome_group.tsv\", index_col=\"genome_id\").genome_group_id\n",
    "    == \"xjin\"\n",
    ")\n",
    "len(genome_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = pd.read_table(\"meta/genome.tsv\", index_col=\"genome_id\").loc[genome_list]\n",
    "assert genome.index.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = list(\n",
    "    genome.loc[genome_list][\n",
    "        lambda x: (x.species_id != \"TODO\") & (x.genome_path != \"\")\n",
    "    ].species_id.unique()\n",
    ")\n",
    "len(species_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"fig/xjin_benchmark_species_list.txt\", \"w\") as f:\n",
    "    for species in species_list:\n",
    "        print(species, file=f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_taxonomy_string(taxonomy_string):\n",
    "    values = taxonomy_string.split(\";\")\n",
    "    return pd.Series(values, index=[\"d__\", \"p__\", \"c__\", \"o__\", \"f__\", \"g__\", \"s__\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy_inpath = \"ref/uhgg_genomes_all_v2.tsv\"\n",
    "\n",
    "species_taxonomy = (\n",
    "    pd.read_table(species_taxonomy_inpath)[lambda x: x.Genome == x.Species_rep]\n",
    "    .assign(species_id=lambda x: \"1\" + x.MGnify_accession.str.split(\"-\").str[2])\n",
    "    .set_index(\"species_id\")\n",
    "    .Lineage.apply(parse_taxonomy_string)\n",
    ")\n",
    "species_taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy.loc[species_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_order = [\n",
    "    \"p__Euryarchaeota\",\n",
    "    \"p__Thermoplasmatota\",\n",
    "    \"p__Firmicutes\",\n",
    "    \"p__Firmicutes_A\",\n",
    "    \"p__Firmicutes_C\",\n",
    "    # \"p__Firmicutes_B\", # None in species_list1\n",
    "    # \"p__Firmicutes_G\", # B/G/I not sure how related to C or A\n",
    "    # \"p__Firmicutes_I\", #\n",
    "    # \"p__Cyanobacteria\", # None in species_list1\n",
    "    \"p__Actinobacteriota\",\n",
    "    \"p__Synergistota\",\n",
    "    \"p__Fusobacteriota\",\n",
    "    \"p__Campylobacterota\",\n",
    "    \"p__Proteobacteria\",\n",
    "    \"p__Desulfobacterota_A\",\n",
    "    \"p__Bacteroidota\",\n",
    "    \"p__Verrucomicrobiota\",\n",
    "    # \"dummy0\", # 18\n",
    "    # \"dummy1\", # 19\n",
    "    # \"dummy2\", # 20\n",
    "]\n",
    "\n",
    "phylum_palette = lib.plot.construct_ordered_palette(\n",
    "    phylum_order,\n",
    "    cm=\"rainbow\",\n",
    "    desaturate_levels=[1.0, 0.5],\n",
    ")\n",
    "\n",
    "for p__ in phylum_order:\n",
    "    print(p__, phylum_palette[p__])\n",
    "    plt.scatter([], [], color=phylum_palette[p__], label=p__)\n",
    "plt.legend(ncols=1)\n",
    "lib.plot.hide_axes_and_spines()\n",
    "\n",
    "# assert len(set(phylum_palette.values())) == len((phylum_palette.values()))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "xjin_sample_list = pd.read_table(\n",
    "    \"meta/XJIN_BENCHMARK/mgen.tsv\", index_col=\"mgen_id\"\n",
    ").index.to_list()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "_depth = pd.read_table(depth_path, names=['sample_id', 'depth'], index_col='sample_id').depth.loc[xjin_sample_list]\n",
    "_depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_group = (\n",
    "    pd.read_table(\"meta/species_group.tsv\")[lambda x: x.species_group_id == \"xjin\"]\n",
    "    .species_id.astype(str)\n",
    "    .to_list()\n",
    ")\n",
    "species_group[:5]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "with open('test.args', 'w') as f:\n",
    "    for genome_id, d in genome.loc[genome_list].iterrows():\n",
    "        species = d.species_id\n",
    "        if species in species_group:\n",
    "            print(f\"data/group/XJIN_BENCHMARK/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc_specgene-ref-t25-p95.{genome_id}.geno_matching_stats.tsv\", file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome[lambda x: x.species_id == \"100003\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_code_meaning = {\n",
    "    -1: \"fail\",\n",
    "    0: \"passes\",\n",
    "    1: \"noise\",\n",
    "    2: \"species-gene\",\n",
    "    3: \"both\",\n",
    "}\n",
    "\n",
    "\n",
    "def assign_qc_code(x):\n",
    "    if x.passes_species_gene_frac:\n",
    "        if x.passes_log_selected_gene_depth_ratio_std:\n",
    "            return 0  # Nothing wrong\n",
    "        else:\n",
    "            return 1  # Doesn't pass noise threshold\n",
    "    elif x.passes_log_selected_gene_depth_ratio_std:\n",
    "        return 2  # Doesn't pass species_gene_frac\n",
    "    else:\n",
    "        return 3  # Doesn't pass either"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data/group/xjin/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc_specgene-ref-t25-p95.STRAIN_MATCH_BENCHMARK_GRID.flag\n",
    "\n",
    "strain_match = {}\n",
    "for genome_id, d in genome.loc[genome_list].iterrows():\n",
    "    species = d.species_id\n",
    "    if species == \"TODO\":\n",
    "        continue\n",
    "    strain_match_path = f\"data/group/xjin/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.{genome_id}.geno_matching_stats.tsv\"\n",
    "    strain_meta_path = f\"data/group/xjin/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.strain_meta-s95-d100-a0-pos100-std25.tsv\"\n",
    "    if os.path.exists(strain_match_path):\n",
    "        strain_meta = pd.read_table(\n",
    "            strain_meta_path, index_col=\"genome_id\"\n",
    "        ).rename_axis(\"strain\")\n",
    "        strain_match[genome_id] = (\n",
    "            pd.read_table(strain_match_path, index_col=[\"genome_id\", \"strain\"])\n",
    "            .assign(\n",
    "                genotype_matching_positions=lambda x: (1 - x.genotype_dissimilarity)\n",
    "                * x.num_geno_positions_compared,\n",
    "                genotype_dissimilarity_pc=lambda x: x.genotype_dissimilarity\n",
    "                + (1 / x.num_geno_positions_compared),\n",
    "                strain_match_path=strain_match_path,\n",
    "            )\n",
    "            .join(strain_meta, on=\"strain\")\n",
    "        )\n",
    "strain_match = (\n",
    "    pd.concat(strain_match.values()).reset_index().set_index([\"genome_id\", \"strain\"])\n",
    ").assign(qc_code=lambda x: x.apply(assign_qc_code, axis=1))\n",
    "strain_match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: We match strains only from among those with \"strain-pure-samples\" in the xjin set\n",
    "# and then picking the lowest genotype_dissimilarity.\n",
    "# TODO: Consider also allowing strains to have strain pure samples\n",
    "# outside the xjin set, but also be detected at appreciable relative abundances\n",
    "# inside the xjin samples?\n",
    "genome_to_spgc_strain = (\n",
    "    strain_match[lambda x: x.strain_depth_max > 0]\n",
    "    .reset_index()\n",
    "    .sort_values(\n",
    "        [\"genotype_dissimilarity_pc\", \"strain_depth_sum\"], ascending=(True, False)\n",
    "    )\n",
    "    .groupby(\"genome_id\")\n",
    "    .head(1)\n",
    "    .set_index(\"genome_id\")\n",
    "    .reindex(genome_list)\n",
    "    .fillna(\n",
    "        {\n",
    "            \"strain\": -1,\n",
    "            \"genotype_dissimilarity\": 1.0,\n",
    "            \"strain_depth_sum\": 0,\n",
    "            \"strain_depth_max\": 0,\n",
    "            \"passes_total_depth\": False,\n",
    "            \"passes_species_gene_frac\": False,\n",
    "            \"passes_gene_count\": False,\n",
    "            \"passes_log_selected_gene_depth_ratio_std\": False,\n",
    "            \"passes_geno_positions\": False,\n",
    "            \"passes_filter\": False,\n",
    "            \"qc_code\": -1,\n",
    "        }\n",
    "    )\n",
    "    .astype({\"strain\": int, \"qc_code\": int})\n",
    ")\n",
    "\n",
    "plt.hist(\n",
    "    genome_to_spgc_strain.genotype_dissimilarity_pc, bins=[0] + list(np.logspace(-5, 0))\n",
    ")\n",
    "plt.xscale(\"symlog\", linthresh=1e-5, linscale=0.1)\n",
    "genome_to_spgc_strain.sort_values(\"genotype_matching_positions\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(\n",
    "    \"strain_depth_sum\",\n",
    "    \"genotype_dissimilarity_pc\",\n",
    "    data=genome_to_spgc_strain,\n",
    "    alpha=0.4,\n",
    ")\n",
    "\n",
    "plt.yscale(\"symlog\", linthresh=1e-4)\n",
    "plt.xscale(\"symlog\", linthresh=1e-2, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data/group/xjin/r.proc.gene99_v15-v22-agg75.spgc_specgene-ref-t25-p95.SPECIES_DEPTH_BENCHMARK_GRID.flag\n",
    "# data/group/xjin/r.proc.gene99_v15-v22-agg75.ACCURACY_BENCHMARK_GRID.flag\n",
    "# data/group/xjin/r.proc.gene99_v15-v22-agg75.spgc-fit.STRAIN_META_BENCHMARK_GRID.flag\n",
    "\n",
    "\n",
    "benchmark = {}\n",
    "missing = []\n",
    "unmatched = []\n",
    "depth_meta = {}\n",
    "spgc_qc = {}\n",
    "for genome_id, d in genome.iterrows():\n",
    "    species = d.species_id\n",
    "    if species == \"TODO\":\n",
    "        continue\n",
    "    depth_path = f\"data/group/xjin/species/sp-{species}/r.proc.gene99_v15-v22-agg75.spgc_specgene-ref-t25-p95.species_depth.tsv\"\n",
    "    _depth = pd.read_table(\n",
    "        depth_path, names=[\"sample_id\", \"depth\"], index_col=\"sample_id\"\n",
    "    ).depth\n",
    "    depth_meta[species] = pd.Series(\n",
    "        dict(species_depth_max=_depth.max(), species_depth_sum=_depth.sum())\n",
    "    )\n",
    "\n",
    "    # SPGC\n",
    "    for unit in [\"uhggtop\", \"eggnog\", \"cog\"]:\n",
    "        for tool in [\n",
    "            \"spgc-fit\",\n",
    "            # \"spgc2-fit\",\n",
    "            # \"nnmatched-m50\",\n",
    "            # \"nnmatched-m10\",\n",
    "            # \"nnmatched-m1\",\n",
    "            # \"nnmatched-m0\",\n",
    "            \"spgc-depth200\",\n",
    "        ]:\n",
    "            accuracy_path = f\"data/group/xjin/species/sp-{species}/r.proc.gene99_v15-v22-agg75.{tool}.{genome_id}.{unit}-reconstruction_accuracy.tsv\"\n",
    "            if not os.path.exists(accuracy_path):\n",
    "                missing.append(accuracy_path)\n",
    "                continue\n",
    "            matched_strain = genome_to_spgc_strain.strain[genome_id]\n",
    "            data = (\n",
    "                pd.read_table(accuracy_path, index_col=\"strain\")\n",
    "                .assign(\n",
    "                    species=species,\n",
    "                    accuracy_path=accuracy_path,\n",
    "                    strain=lambda x: x.index,\n",
    "                )\n",
    "                .sort_values(\"f1\", ascending=False)\n",
    "            )\n",
    "\n",
    "            if matched_strain in data.index:\n",
    "                benchmark[(genome_id, tool, unit, \"match\")] = data.loc[matched_strain]\n",
    "            if data.shape[0] >= 1:\n",
    "                benchmark[(genome_id, tool, unit, \"top\")] = data.iloc[0]\n",
    "            if data.shape[0] >= 2:\n",
    "                benchmark[(genome_id, tool, unit, \"second\")] = data.iloc[1]\n",
    "        for tool in [\n",
    "            \"panphlan\",\n",
    "            # \"spanda-s2\",\n",
    "            \"spanda-s3\",\n",
    "            \"spanda-s4\",\n",
    "            \"spanda-s5\",\n",
    "            \"spanda-s6\",\n",
    "        ]:\n",
    "            accuracy_path = f\"data/group/xjin/species/sp-{species}/r.proc.gene99_v15-v22-agg75.{tool}.{genome_id}.{unit}-reconstruction_accuracy.tsv\"\n",
    "            if not os.path.exists(accuracy_path):\n",
    "                missing.append(accuracy_path)\n",
    "                continue\n",
    "            data = (\n",
    "                pd.read_table(accuracy_path, index_col=\"strain\")\n",
    "                .assign(\n",
    "                    species=species,\n",
    "                    accuracy_path=accuracy_path,\n",
    "                    strain=lambda x: x.index,\n",
    "                )\n",
    "                .sort_values(\"f1\", ascending=False)\n",
    "            )\n",
    "            if data.shape[0] >= 1:\n",
    "                benchmark[(genome_id, tool, unit, \"top\")] = data.iloc[0]\n",
    "                benchmark[(genome_id, tool, unit, \"match\")] = data.iloc[0]\n",
    "            if data.shape[0] >= 2:\n",
    "                benchmark[(genome_id, tool, unit, \"second\")] = data.iloc[1]\n",
    "            if data.shape[0] >= 3:\n",
    "                benchmark[(genome_id, tool, unit, \"third\")] = data.iloc[2]\n",
    "\n",
    "    spgc_qc_path = f\"data/group/xjin/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.strain_meta-s95-d100-a0-pos100-std25.tsv\"\n",
    "    if not os.path.exists(spgc_qc_path):\n",
    "        continue\n",
    "    _qc = pd.read_table(spgc_qc_path, index_col=\"genome_id\")\n",
    "    if matched_strain in _qc.index:\n",
    "        spgc_qc[genome_id] = _qc.loc[matched_strain]\n",
    "\n",
    "benchmark = pd.DataFrame(benchmark.values(), index=benchmark.keys()).rename_axis(\n",
    "    [\"genome_id\", \"tool\", \"unit\", \"match\"]\n",
    ")\n",
    "depth_meta = pd.DataFrame(depth_meta.values(), index=depth_meta.keys()).rename_axis(\n",
    "    \"species\"\n",
    ")\n",
    "spgc_qc = pd.DataFrame(spgc_qc.values(), index=spgc_qc.keys()).assign(\n",
    "    qc_code=lambda x: x.apply(assign_qc_code, axis=1)\n",
    ")\n",
    "\n",
    "benchmark.sort_values([\"species\", \"tool\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many additional species had _no_ depth whatsoever?\n",
    "\n",
    "genome.join(depth_meta, on=\"species_id\")[lambda x: x.species_depth_sum == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# And how many had any depth?\n",
    "\n",
    "genome.join(depth_meta, on=\"species_id\")[lambda x: x.species_depth_max.fillna(0) > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_filt_list = list(\n",
    "    genome.join(depth_meta, on=\"species_id\")[\n",
    "        lambda x: (x.species_id != \"TODO\") & (x.species_depth_sum > 0)\n",
    "    ].index\n",
    ")\n",
    "species_filt_list = list(genome.loc[genome_filt_list].species_id.unique())\n",
    "\n",
    "len(genome_filt_list), len(species_filt_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many actual SPGC fits are there?\n",
    "\n",
    "benchmark.xs(\n",
    "    [\"spgc-fit\", \"eggnog\", \"match\"], level=(\"tool\", \"unit\", \"match\")\n",
    ").f1.sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which theoretically estimate-able genomes have no StrainPGC results?\n",
    "\n",
    "set(genome_filt_list) - set(\n",
    "    benchmark.xs([\"spgc-fit\", \"eggnog\", \"match\"], level=(\"tool\", \"unit\", \"match\")).index\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What's wrong with this strain?\n",
    "genome.loc[\"Collinsella-aerofaciens-ATCC-25986_MAF-2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which zero-depth species have StrainPanDA results?\n",
    "\n",
    "set(\n",
    "    genome.join(depth_meta, on=\"species_id\")[\n",
    "        lambda x: x.species_depth_max.fillna(0) == 0\n",
    "    ].index\n",
    ") & set(\n",
    "    benchmark.xs(\n",
    "        [\"spanda-s5\", \"eggnog\", \"match\"], level=(\"tool\", \"unit\", \"match\")\n",
    "    ).index\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which zero-depth species have PanPhlAn results?\n",
    "\n",
    "set(\n",
    "    genome.join(depth_meta, on=\"species_id\")[\n",
    "        lambda x: x.species_depth_max.fillna(0) == 0\n",
    "    ].index\n",
    ") & set(\n",
    "    benchmark.xs([\"panphlan\", \"eggnog\", \"match\"], level=(\"tool\", \"unit\", \"match\")).index\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How did the other tools perform on these?\n",
    "benchmark.xs([\"eggnog\", \"match\"], level=(\"unit\", \"match\")).loc[\n",
    "    [\"Roseburia-inulinivorans-DSM-16841_MAF-2\"]\n",
    "][[\"precision\", \"recall\", \"f1\"]].unstack(\"tool\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "benchmark.xs('panphlan', level='tool').accuracy_path.iloc[0]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "benchmark.xs(('panphlan', 'eggnog', 'match'), level=('tool', 'unit', 'match')).sort_values('f1').accuracy_path.values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=50)\n",
    "d = benchmark.xs((\"eggnog\", \"top\"), level=(\"unit\", \"match\")).f1.unstack().fillna(0)\n",
    "for tool in [\n",
    "    \"spgc-fit\",\n",
    "    # \"nnmatched-m0\", \"nnmatched-m50\",\n",
    "    \"spgc-depth200\",\n",
    "    # \"spanda-s4\",\n",
    "]:\n",
    "    plt.hist(d[tool], bins=bins, alpha=0.5, label=tool)\n",
    "# plt.hist(d[\"nnmatched-m0\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"nnmatched-m50\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spgc-depth200\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"panphlan\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spanda-s2\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spanda-s3\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spanda-s4\"], bins=bins, alpha=0.5)\n",
    "plt.legend()\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=50)\n",
    "d = benchmark.xs((\"eggnog\", \"top\"), level=(\"unit\", \"match\")).f1.unstack().fillna(0)\n",
    "for tool in [\n",
    "    # \"spgc-fit\",\n",
    "    # \"nnmatched-m0\", \"nnmatched-m50\",\n",
    "    # \"spgc-depth200\",\n",
    "    # \"spanda-s2\",\n",
    "    \"spanda-s3\",\n",
    "    \"spanda-s4\",\n",
    "    \"spanda-s5\",\n",
    "]:\n",
    "    plt.hist(d[tool], bins=bins, alpha=0.5, label=tool)\n",
    "# plt.hist(d[\"nnmatched-m0\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"nnmatched-m50\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spgc-depth200\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"panphlan\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spanda-s2\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spanda-s3\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spanda-s4\"], bins=bins, alpha=0.5)\n",
    "plt.legend()\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = genome_to_spgc_strain.join(\n",
    "    benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack()\n",
    ").fillna(\n",
    "    0\n",
    ")  # .loc[genome_list]\n",
    "\n",
    "plt.scatter(\n",
    "    \"spanda-s5\",\n",
    "    \"spgc-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    c=\"strain_depth_max\",\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = genome_to_spgc_strain.join(\n",
    "    benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack()\n",
    ").fillna(\n",
    "    0\n",
    ")  # .loc[genome_list]\n",
    "\n",
    "plt.scatter(\n",
    "    \"spanda-s4\",\n",
    "    \"spanda-s6\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    c=\"strain_depth_max\",\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark.xs(\n",
    "    (\"eggnog\", \"match\", \"spanda-s5\"), level=(\"unit\", \"match\", \"tool\")\n",
    ").sort_values(\"f1\", ascending=False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "benchmark.xs(('eggnog', 'spgc-fit', 'match'), level=('unit', 'tool', 'match')).accuracy_path.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = genome_to_spgc_strain.join(\n",
    "    benchmark.xs((\"spgc-fit\", \"eggnog\"), level=(\"tool\", \"unit\")).f1.unstack()\n",
    ").fillna(0)\n",
    "\n",
    "\n",
    "plt.scatter(\n",
    "    \"match\",\n",
    "    \"top\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    # c=\"strain_depth_max\",\n",
    ")\n",
    "# plt.colorbar()\n",
    "# # plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = genome_to_spgc_strain.join(\n",
    "    benchmark.xs((\"match\", \"eggnog\"), level=(\"match\", \"unit\")).f1.unstack()\n",
    ").fillna(0)\n",
    "\n",
    "\n",
    "plt.scatter(\n",
    "    \"spgc-fit\",\n",
    "    \"spgc2-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    # c=\"strain_depth_max\",\n",
    ")\n",
    "# plt.colorbar()\n",
    "# # plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(\n",
    "    \"precision\",\n",
    "    \"recall\",\n",
    "    data=benchmark.xs((\"spgc-fit\", \"eggnog\", \"match\"), level=(\"tool\", \"unit\", \"match\")),\n",
    ")\n",
    "plt.plot([0, 1], [0, 1])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0[lambda x: x['spgc-fit'] != x['spgc2-fit']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which genomes are failing and why?\n",
    "(\n",
    "    genome_to_spgc_strain.join(\n",
    "        benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack()\n",
    "    )\n",
    "    .fillna(0)[lambda x: x[\"spgc-fit\"] == 0]\n",
    "    .join(genome)\n",
    "    # .join(depth_meta, on=\"species_id\")\n",
    "    .join(\n",
    "        genome.species_id.value_counts().rename(\"num_strains_in_species\"),\n",
    "        on=\"species_id\",\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What's the point of benchmarking strains that have no or very low depth?\n",
    "# that just even more heavily favors the tools that I'm taking \"best hit\" for.\n",
    "d0 = (\n",
    "    genome_to_spgc_strain.join(\n",
    "        benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack()\n",
    "    )\n",
    "    .fillna(0)\n",
    "    .join(genome)\n",
    "    # .join(depth_meta, on=\"species_id\")\n",
    "    [lambda x: x.species_depth_sum >= 0.5]\n",
    ")\n",
    "print(len(d0))\n",
    "\n",
    "plt.scatter(\n",
    "    \"panphlan\",\n",
    "    \"spgc-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    c=\"strain_depth_max\",\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    genome_to_spgc_strain.join(\n",
    "        benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack()\n",
    "    )\n",
    "    .fillna(0)\n",
    "    .join(genome)\n",
    "    # .join(depth_meta, on=\"species_id\")\n",
    "    [lambda x: x.species_depth_sum >= 0.5]\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    \"panphlan\",\n",
    "    \"spgc-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    c=\"strain_depth_max\",\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.ylim(0.55, 1.05)\n",
    "plt.xlim(0.55, 1.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\"))\n",
    "    .f1.unstack()\n",
    "    .fillna(0)\n",
    "    .join(\n",
    "        genome_to_spgc_strain[\n",
    "            [\n",
    "                \"genotype_dissimilarity\",\n",
    "                \"strain_depth_sum\",\n",
    "                \"strain_depth_max\",\n",
    "            ]\n",
    "        ]\n",
    "    )\n",
    "    .fillna({\"genotype_dissimilarity\": 1.0})\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    \"genotype_dissimilarity\",\n",
    "    \"spgc-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    c=\"strain_depth_max\",\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.1),\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.xscale(\"symlog\", linthresh=1e-5, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\"))\n",
    "    .f1.unstack()\n",
    "    .fillna(0)\n",
    "    .join(\n",
    "        genome_to_spgc_strain[\n",
    "            [\n",
    "                \"genotype_dissimilarity\",\n",
    "                \"strain_depth_sum\",\n",
    "                \"strain_depth_max\",\n",
    "            ]\n",
    "        ]\n",
    "    )\n",
    "    .fillna({\"genotype_dissimilarity\": 1.0})\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    \"strain_depth_sum\",\n",
    "    \"genotype_dissimilarity\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    c=\"spgc-fit\",\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-4)\n",
    "plt.xscale(\"symlog\", linthresh=1e-2, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\"))\n",
    "    .f1.unstack()\n",
    "    .fillna(0)\n",
    "    .join(\n",
    "        genome_to_spgc_strain[\n",
    "            [\n",
    "                \"genotype_dissimilarity\",\n",
    "                \"strain_depth_sum\",\n",
    "                \"strain_depth_max\",\n",
    "            ]\n",
    "        ]\n",
    "    )\n",
    "    .fillna({\"genotype_dissimilarity\": 1.0})\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    \"strain_depth_sum\",\n",
    "    \"genotype_dissimilarity\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    c=\"spgc-fit\",\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-4)\n",
    "plt.xscale(\"symlog\", linthresh=1e-2, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multi_genome_species = idxwhere(\n",
    "    strain_match.join(genome.species_id)\n",
    "    .reset_index()[[\"species_id\", \"genome_id\"]]\n",
    "    .drop_duplicates()\n",
    "    .species_id.value_counts()\n",
    "    > 1\n",
    ")\n",
    "multi_genome_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    genome.join(genome_to_spgc_strain)[\n",
    "        lambda x: (x.species_id != \"TODO\")\n",
    "        & (x.species_depth_sum > 0)\n",
    "        & (x.species_depth_max > 0)\n",
    "        # & ~x.species_id.isin(multi_genome_species)\n",
    "    ]\n",
    "    .join(\n",
    "        benchmark.xs((\"spgc-fit\", \"eggnog\", \"match\"), level=(\"tool\", \"unit\", \"match\")),\n",
    "        rsuffix=\"_\",\n",
    "    )\n",
    "    .fillna({\"precision\": 0, \"recall\": 0, \"f1\": 0, \"jaccard\": 0})\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How many strains/species/phyla are analyzed in the benchmark?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Full list\n",
    "\n",
    "len(genome_list), len(species_list), len(\n",
    "    species_taxonomy.loc[species_list].p__.unique()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtered list\n",
    "len(genome_filt_list), len(species_filt_list), len(\n",
    "    species_taxonomy.loc[species_filt_list].p__.unique()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Which species were excluded?\n",
    "#  - For too few GT-Pro positions?\n",
    "#  - For too little depth?\n",
    "#  - For other reasons???\n",
    "#  - What was supposed to be included but basically failed, and why?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## StrainPGC Performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tool = \"spgc-fit\"\n",
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "d = (\n",
    "    genome_to_spgc_strain.loc[genome_filt_list]\n",
    "    .join(benchmark.xs((unit, match), level=(\"unit\", \"match\")), rsuffix=\"_\")[\n",
    "        [\"precision\", \"recall\", \"f1\"]\n",
    "    ]\n",
    "    .unstack(\"tool\")\n",
    "    .xs(tool, level=\"tool\", axis=\"columns\")\n",
    "    .fillna(0)\n",
    "    .join(genome_to_spgc_strain.species_gene_frac)\n",
    ")\n",
    "\n",
    "# plt.scatter('precision', \"recall\", data=d, c='species_gene_frac')\n",
    "# plt.colorbar()\n",
    "# plt.plot([0, 1], [0, 1])\n",
    "print(d[[\"precision\", \"recall\", \"f1\"]].quantile([0.25, 0.5, 0.75]))\n",
    "# print(d[d.species_gene_frac > 0.9][[\"precision\", \"recall\", \"f1\", \"jaccard\"]].quantile([0.25, 0.5, 0.75]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How does it compare to other tools?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    genome_to_spgc_strain.loc[genome_filt_list]\n",
    "    .join(benchmark.xs((unit, match), level=(\"unit\", \"match\")), rsuffix=\"_\")\n",
    "    .f1.unstack(\"tool\")\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "\n",
    "print((-(d.subtract(d[\"spgc-fit\"], axis=0))).quantile([0.25, 0.5, 0.75]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "qc_code_palette = {\n",
    "    0: \"grey\",\n",
    "    1: \"lightgreen\",\n",
    "    2: \"lightblue\",\n",
    "    3: \"lightsalmon\",\n",
    "    -1: \"white\",\n",
    "}\n",
    "\n",
    "d0 = benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "\n",
    "_tool_comparison_order = [\n",
    "    \"panphlan\",\n",
    "    \"spanda-s6\",\n",
    "    # \"spgc-depth200\",\n",
    "]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2 * len(_tool_comparison_order), 2 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y_tool = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d1 = (\n",
    "        d0[score]\n",
    "        .unstack(fill_value=0)\n",
    "        .join(genome_to_spgc_strain)\n",
    "        .assign(\n",
    "            c=lambda x: x.qc_code.astype(int).map(\n",
    "                qc_code_palette,\n",
    "            )\n",
    "        )\n",
    "        .sort_values(\"qc_code\")\n",
    "    )\n",
    "    print(\n",
    "        score, \"SPGC IQR:\", (d1[y_tool]).quantile([0.25, 0.5, 0.75]).round(3).tolist()\n",
    "    )\n",
    "    for x_tool, ax in zip(_tool_comparison_order, ax_row):\n",
    "        print(\n",
    "            \"compared to: {}, {}, {:.1g}\".format(\n",
    "                x_tool,\n",
    "                (d1[y_tool] - d1[x_tool])\n",
    "                .quantile([0.05, 0.25, 0.5, 0.75, 0.95])\n",
    "                .round(3)\n",
    "                .tolist(),\n",
    "                sp.stats.wilcoxon(d1[x_tool], d1[y_tool]).pvalue,\n",
    "            )\n",
    "        )\n",
    "        ax.scatter(\n",
    "            x_tool,\n",
    "            y_tool,\n",
    "            data=d1,\n",
    "            c=\"c\",\n",
    "            lw=0.5,\n",
    "            edgecolor=\"k\",\n",
    "            s=15,\n",
    "            label=\"__nolegend__\",\n",
    "        )\n",
    "        ax.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "    print()\n",
    "\n",
    "for _score_order, ax in zip(_score_order, axs[:, 0]):\n",
    "    ax.set_ylabel(_score_order)\n",
    "    ax.set_yticks([0, 0.5, 1])\n",
    "    ax.set_ylim(-0.05, 1.05)\n",
    "\n",
    "for x_tool, ax in zip(_tool_comparison_order, axs[-1, :]):\n",
    "    ax.set_xlabel(x_tool)\n",
    "    ax.set_xticks([0, 0.5, 1])\n",
    "    ax.set_xlim(-0.05, 1.05)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(2, 1.25))\n",
    "for i in range(-1, 4):\n",
    "    ax.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        label=qc_code_meaning[i],\n",
    "        color=qc_code_palette[i],\n",
    "        lw=0.5,\n",
    "        edgecolor=\"k\",\n",
    "        s=15,\n",
    "    )\n",
    "ax.legend(markerscale=2, loc=\"lower right\")\n",
    "lib.plot.hide_axes_and_spines(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "qc_code_palette = {\n",
    "    0: \"grey\",\n",
    "    1: \"lightgreen\",\n",
    "    2: \"lightblue\",\n",
    "    3: \"lightsalmon\",\n",
    "    -1: \"white\",\n",
    "}\n",
    "\n",
    "d0 = benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "\n",
    "_tool_comparison_order = [\n",
    "    \"panphlan\",\n",
    "    \"spanda-s6\",\n",
    "    # \"spgc-depth200\",\n",
    "]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2 * len(_tool_comparison_order), 2 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    "    gridspec_kw=dict(hspace=0.1, wspace=0.1),\n",
    ")\n",
    "y_tool = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d1 = (\n",
    "        d0[score]\n",
    "        .unstack(fill_value=0)\n",
    "        .join(genome_to_spgc_strain)\n",
    "        .assign(\n",
    "            c=lambda x: x.qc_code.astype(int).map(\n",
    "                qc_code_palette,\n",
    "            )\n",
    "        )\n",
    "        .sort_values(\"qc_code\")\n",
    "    )\n",
    "    print(\n",
    "        score, \"SPGC IQR:\", (d1[y_tool]).quantile([0.25, 0.5, 0.75]).round(3).tolist()\n",
    "    )\n",
    "    for x_tool, ax in zip(_tool_comparison_order, ax_row):\n",
    "        print(\n",
    "            \"compared to: {}, {}, {:.1g}\".format(\n",
    "                x_tool,\n",
    "                (d1[y_tool] - d1[x_tool])\n",
    "                .quantile([0.05, 0.25, 0.5, 0.75, 0.95])\n",
    "                .round(3)\n",
    "                .tolist(),\n",
    "                sp.stats.wilcoxon(d1[x_tool], d1[y_tool]).pvalue,\n",
    "            )\n",
    "        )\n",
    "        ax.scatter(\n",
    "            x_tool,\n",
    "            y_tool,\n",
    "            data=d1,\n",
    "            # c=\"c\",\n",
    "            color='grey',\n",
    "            lw=0.25,\n",
    "            edgecolor=\"k\",\n",
    "            s=15,\n",
    "            label=\"__nolegend__\",\n",
    "        )\n",
    "        ax.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "    print()\n",
    "\n",
    "for _score_order, ax in zip(_score_order, axs[:, 0]):\n",
    "    ax.set_ylabel(_score_order)\n",
    "    ax.set_yticks([0, 0.5, 1])\n",
    "    ax.set_ylim(-0.05, 1.05)\n",
    "\n",
    "for x_tool, ax in zip(_tool_comparison_order, axs[-1, :]):\n",
    "    ax.set_xlabel(x_tool)\n",
    "    ax.set_xticks([0, 0.5, 1])\n",
    "    ax.set_xlim(-0.05, 1.05)\n",
    "\n",
    "# fig, ax = plt.subplots(figsize=(2, 1.25))\n",
    "# for i in range(-1, 4):\n",
    "#     ax.scatter(\n",
    "#         [],\n",
    "#         [],\n",
    "#         label=qc_code_meaning[i],\n",
    "#         color=qc_code_palette[i],\n",
    "#         lw=0.5,\n",
    "#         edgecolor=\"k\",\n",
    "#         s=15,\n",
    "#     )\n",
    "# ax.legend(markerscale=2, loc=\"lower right\")\n",
    "# lib.plot.hide_axes_and_spines(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "qc_code_palette = {\n",
    "    0: \"grey\",\n",
    "    1: \"lightgreen\",\n",
    "    2: \"lightblue\",\n",
    "    3: \"lightsalmon\",\n",
    "    -1: \"grey\",\n",
    "}\n",
    "\n",
    "d0 = benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "\n",
    "_tool_comparison_order = [\n",
    "    \"panphlan\",\n",
    "    \"spanda-s6\",\n",
    "    # \"spgc-depth200\",\n",
    "]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2 * len(_tool_comparison_order), 2 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    "    gridspec_kw=dict(hspace=0.1, wspace=0.1),\n",
    ")\n",
    "y_tool = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d1 = (\n",
    "        d0[score]\n",
    "        .unstack(fill_value=0)\n",
    "        .join(genome_to_spgc_strain)\n",
    "        .assign(\n",
    "            c=lambda x: x.qc_code.astype(int).map(\n",
    "                qc_code_palette,\n",
    "            )\n",
    "        )\n",
    "        .sort_values(\"qc_code\")\n",
    "    )\n",
    "    print(\n",
    "        score, \"SPGC IQR:\", (d1[y_tool]).quantile([0.25, 0.5, 0.75]).round(3).tolist()\n",
    "    )\n",
    "    for x_tool, ax in zip(_tool_comparison_order, ax_row):\n",
    "        print(\n",
    "            \"compared to: {}, {}, {:.1g}\".format(\n",
    "                x_tool,\n",
    "                (d1[y_tool] - d1[x_tool])\n",
    "                .quantile([0.05, 0.25, 0.5, 0.75, 0.95])\n",
    "                .round(3)\n",
    "                .tolist(),\n",
    "                sp.stats.wilcoxon(d1[x_tool], d1[y_tool]).pvalue,\n",
    "            )\n",
    "        )\n",
    "        ax.scatter(\n",
    "            x_tool,\n",
    "            y_tool,\n",
    "            data=d1,\n",
    "            # c=\"c\",\n",
    "            lw=.8,\n",
    "            edgecolors=\"c\",\n",
    "            facecolor='none',\n",
    "            s=30,\n",
    "            alpha=0.75,\n",
    "            label=\"__nolegend__\",\n",
    "        )\n",
    "        ax.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "    print()\n",
    "\n",
    "for _score_order, ax in zip(_score_order, axs[:, 0]):\n",
    "    ax.set_ylabel(_score_order)\n",
    "    ax.set_yticks([0, 0.5, 1])\n",
    "    ax.set_ylim(-0.05, 1.05)\n",
    "\n",
    "for x_tool, ax in zip(_tool_comparison_order, axs[-1, :]):\n",
    "    ax.set_xlabel(x_tool)\n",
    "    ax.set_xticks([0, 0.5, 1])\n",
    "    ax.set_xlim(-0.05, 1.05)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(2, 1.25))\n",
    "for i in range(-1, 4):\n",
    "    ax.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        label=qc_code_meaning[i],\n",
    "        # color=qc_code_palette[i],\n",
    "        lw=1,\n",
    "        edgecolor=qc_code_palette[i],\n",
    "        facecolor='none',\n",
    "        s=30,\n",
    "    )\n",
    "ax.legend(markerscale=2, loc=\"lower right\")\n",
    "lib.plot.hide_axes_and_spines(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "# qc_code_palette = {\n",
    "#     0: \"grey\",\n",
    "#     1: \"lightgreen\",\n",
    "#     2: \"lightblue\",\n",
    "#     3: \"lightsalmon\",\n",
    "#     -1: \"grey\",\n",
    "# }\n",
    "\n",
    "cmap='Greys'\n",
    "norm=mpl.colors.LogNorm(vmin=0.2, vmax=50)\n",
    "\n",
    "d0 = benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "\n",
    "_tool_comparison_order = [\n",
    "    \"panphlan\",\n",
    "    \"spanda-s6\",\n",
    "    # \"spgc-depth200\",\n",
    "]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2 * len(_tool_comparison_order), 2 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    "    gridspec_kw=dict(hspace=0.1, wspace=0.1),\n",
    "    squeeze=False,\n",
    ")\n",
    "y_tool = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d1 = (\n",
    "        d0[score]\n",
    "        .unstack(fill_value=0)\n",
    "        .join(genome_to_spgc_strain)\n",
    "        .assign(\n",
    "            c=lambda x: x.qc_code.astype(int).map(\n",
    "                qc_code_palette,\n",
    "            )\n",
    "        )\n",
    "        .sort_values(\"qc_code\")\n",
    "    )\n",
    "    print(\n",
    "        score, \"SPGC IQR:\", (d1[y_tool]).quantile([0.25, 0.5, 0.75]).round(3).tolist()\n",
    "    )\n",
    "    for x_tool, ax in zip(_tool_comparison_order, ax_row):\n",
    "        print(\n",
    "            \"compared to: {}, {}, {:.1g}\".format(\n",
    "                x_tool,\n",
    "                (d1[y_tool] - d1[x_tool])\n",
    "                .quantile([0.05, 0.25, 0.5, 0.75, 0.95])\n",
    "                .round(3)\n",
    "                .tolist(),\n",
    "                sp.stats.wilcoxon(d1[x_tool], d1[y_tool]).pvalue,\n",
    "            )\n",
    "        )\n",
    "        art = ax.hist2d(\n",
    "            x_tool,\n",
    "            y_tool,\n",
    "            data=d1,\n",
    "            # # c=\"c\",\n",
    "            # lw=1,\n",
    "            # edgecolor=\"grey\",\n",
    "            # facecolor='none',\n",
    "            # s=30,\n",
    "            # alpha=0.75,\n",
    "            # label=\"__nolegend__\",\n",
    "            cmap=cmap,\n",
    "            norm=norm,\n",
    "            bins=np.linspace(0, 1, num=21),\n",
    "        )\n",
    "        ax.plot([0, 1], [0, 1], lw=0.5, linestyle=\"--\", color=\"k\")\n",
    "    print()\n",
    "\n",
    "for _score_order, ax in zip(_score_order, axs[:, 0]):\n",
    "    ax.set_ylabel(_score_order)\n",
    "    ax.set_yticks([0, 0.5, 1])\n",
    "    ax.set_ylim(-0.05, 1.05)\n",
    "\n",
    "for x_tool, ax in zip(_tool_comparison_order, axs[-1, :]):\n",
    "    ax.set_xlabel(x_tool)\n",
    "    ax.set_xticks([0, 0.5, 1])\n",
    "    ax.set_xlim(-0.05, 1.05)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(0.1, 1.25))\n",
    "fig.colorbar(mpl.cm.ScalarMappable(norm=norm, cmap=cmap), cax=ax)\n",
    "ax.set_yscale('log', subs=[])\n",
    "ax.set_yticks([1, 10, 50])\n",
    "# for i in range(-1, 4):\n",
    "#     ax.scatter(\n",
    "#         [],\n",
    "#         [],\n",
    "#         label=qc_code_meaning[i],\n",
    "#         # color=qc_code_palette[i],\n",
    "#         lw=1,\n",
    "#         edgecolor=qc_code_palette[i],\n",
    "#         facecolor='none',\n",
    "#         s=30,\n",
    "#     )\n",
    "# ax.legend(markerscale=2, loc=\"lower right\")\n",
    "# lib.plot.hide_axes_and_spines(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "qc_code_palette = {\n",
    "    0: \"grey\",\n",
    "    1: \"lightgreen\",\n",
    "    2: \"lightsalmon\",\n",
    "    3: \"violet\",\n",
    "    -1: \"white\",\n",
    "}\n",
    "\n",
    "d0 = benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "\n",
    "_tool_comparison_order = [\n",
    "    \"panphlan\",\n",
    "    \"spanda-s4\",\n",
    "    \"spanda-s5\",\n",
    "    \"spgc-fit\",\n",
    "]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(3 * len(_tool_comparison_order), 3 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y_tool = \"spanda-s6\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d1 = (\n",
    "        d0[score]\n",
    "        .unstack(fill_value=0)\n",
    "        .join(genome_to_spgc_strain)\n",
    "        .assign(\n",
    "            c=lambda x: x.qc_code.fillna(-1)\n",
    "            .astype(int)\n",
    "            .map(\n",
    "                qc_code_palette,\n",
    "            )\n",
    "        )\n",
    "        .sort_values(\"qc_code\")\n",
    "    )\n",
    "    print(\n",
    "        score, \"SPGC IQR:\", (d1[y_tool]).quantile([0.25, 0.5, 0.75]).round(3).tolist()\n",
    "    )\n",
    "    for x_tool, ax in zip(_tool_comparison_order, ax_row):\n",
    "        print(\n",
    "            \"compared to: {}, {}, {:.1g}\".format(\n",
    "                x_tool,\n",
    "                (d1[y_tool] - d1[x_tool])\n",
    "                .quantile([0.05, 0.25, 0.5, 0.75, 0.95])\n",
    "                .round(3)\n",
    "                .tolist(),\n",
    "                sp.stats.wilcoxon(d1[x_tool], d1[y_tool]).pvalue,\n",
    "            )\n",
    "        )\n",
    "        ax.scatter(\n",
    "            x_tool,\n",
    "            y_tool,\n",
    "            data=d1,\n",
    "            c=\"c\",\n",
    "            lw=0.5,\n",
    "            edgecolor=\"k\",\n",
    "            s=15,\n",
    "            label=\"__nolegend__\",\n",
    "        )\n",
    "        ax.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "    print()\n",
    "\n",
    "for _score_order, ax in zip(_score_order, axs[:, 0]):\n",
    "    ax.set_ylabel(_score_order)\n",
    "    ax.set_yticks([0, 0.5, 1])\n",
    "    ax.set_ylim(-0.05, 1.05)\n",
    "\n",
    "for x_tool, ax in zip(_tool_comparison_order, axs[-1, :]):\n",
    "    ax.set_xlabel(x_tool)\n",
    "    ax.set_xticks([0, 0.5, 1])\n",
    "    ax.set_xlim(-0.05, 1.05)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(2, 1.25))\n",
    "for i in range(-1, 4):\n",
    "    ax.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        label=qc_code_meaning[i],\n",
    "        color=qc_code_palette[i],\n",
    "        lw=0.5,\n",
    "        edgecolor=\"k\",\n",
    "        s=15,\n",
    "    )\n",
    "ax.legend(markerscale=2, loc=\"lower right\")\n",
    "lib.plot.hide_axes_and_spines(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FIXME: Load data as in the main-text figure.\n",
    "\n",
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "_tool_list = [\n",
    "    \"spgc-fit\",\n",
    "    \"panphlan\",\n",
    "    \"spanda-s6\",\n",
    "    \"spgc-depth200\",\n",
    "    # \"nnmatched-m50\"\n",
    "]\n",
    "_score_list = [\"precision\", \"recall\", \"f1\"]\n",
    "xvar = \"strain_depth_max\"\n",
    "\n",
    "# _tool_palette = lib.plot.construct_ordered_palette(_tool_list)\n",
    "_tool_palette = lib.plot.construct_ordered_palette_from_list(\n",
    "    _tool_list, colors=[\"tab:blue\", \"tab:green\", \"tab:orange\", \"tab:purple\"]\n",
    ")\n",
    "\n",
    "\n",
    "d0 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))[_score_list]\n",
    "    .reindex(product(genome_filt_list, _tool_list))\n",
    "    .fillna(0)\n",
    "    .join(genome)\n",
    "    # [lambda x: ~x.species_id.isin(multi_genome_species)]\n",
    "    .join(depth_meta, on=\"species_id\")\n",
    "    .join(genome_to_spgc_strain, rsuffix=\"_\")\n",
    ")\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_list),\n",
    "    figsize=(5, 3 * len(_score_list)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "\n",
    "for _tool in _tool_list:\n",
    "    d1 = d0.xs(_tool, level=\"tool\")\n",
    "    for _score, ax in zip(_score_list, axs.flatten()):\n",
    "        fit = smf.ols(f'{_score} ~ cr({xvar}, 5, constraints=\"center\")', data=d1).fit()\n",
    "        d2 = d1.assign(\n",
    "            lowess_y=lambda d: sm.nonparametric.lowess(\n",
    "                d[_score], d[xvar], it=10, frac=1 / 3, return_sorted=False\n",
    "            ),\n",
    "            spline_y=lambda d: fit.predict(pd.DataFrame({xvar: d[xvar]})),\n",
    "            xvar_rolling_average=lambda d: d.sort_values(xvar)\n",
    "            .rolling(window=10)[xvar]\n",
    "            .mean(),\n",
    "            score_rolling_average=lambda d: d.sort_values(xvar)\n",
    "            .rolling(window=10)[_score]\n",
    "            .mean(),\n",
    "        )\n",
    "        smoothed = pd.DataFrame({xvar: np.logspace(-0.5, 4, num=100)}).assign(\n",
    "            spline_y=lambda d: fit.predict(pd.DataFrame({xvar: d[xvar]})),\n",
    "            lowess_y=lambda d: sm.nonparametric.lowess(\n",
    "                d2[_score],\n",
    "                d2[xvar],\n",
    "                xvals=d[xvar],\n",
    "                it=10,\n",
    "                frac=1 / 3,\n",
    "                return_sorted=False,\n",
    "            ),\n",
    "        )\n",
    "        ax.scatter(\n",
    "            xvar,\n",
    "            _score,\n",
    "            data=d2,\n",
    "            label=_tool,\n",
    "            s=8,\n",
    "            alpha=0.5,\n",
    "            facecolor=\"none\",\n",
    "            color=_tool_palette[_tool],\n",
    "        )\n",
    "        print(_tool, _score, sp.stats.spearmanr(d2[xvar], d2[_score]))\n",
    "        ax.plot(\n",
    "            \"xvar_rolling_average\",\n",
    "            \"score_rolling_average\",\n",
    "            data=d2.sort_values(xvar),\n",
    "            label=\"__nolegend__\",\n",
    "            lw=1,\n",
    "            linestyle=\"-\",\n",
    "            color=_tool_palette[_tool],\n",
    "        )\n",
    "\n",
    "for _score, ax in zip(_score_list, axs.flatten()):\n",
    "    ax.set_ylabel(_score)\n",
    "\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-1, linscale=0.1)\n",
    "\n",
    "pad = 1e-2\n",
    "# ax.set_yscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "# ax.set_yticks(np.unique([0.0, 0.5, 0.75, 0.9, 0.95, 0.99]))\n",
    "ax.set_ylim(-0.05, 1.05)\n",
    "\n",
    "ax.legend(bbox_to_anchor=(1, 1))\n",
    "\n",
    "print(len(d2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_depth = []\n",
    "_missing_species = []\n",
    "\n",
    "for species in tqdm(species_list):\n",
    "    inpath = f\"data/group/xjin/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.species_depth.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        _missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath, names=[\"sample\", \"depth\"]).assign(species=species)\n",
    "    species_depth.append(data)\n",
    "species_depth = (\n",
    "    pd.concat(species_depth)\n",
    "    .set_index([\"sample\", \"species\"])\n",
    "    .depth.unstack(fill_value=0)\n",
    ")\n",
    "\n",
    "print(\n",
    "    len(_missing_species),\n",
    "    \"out of\",\n",
    "    len(species_list),\n",
    "    \"species are missing.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_depth = []\n",
    "missing_files = []\n",
    "for species_id in species_depth.columns:\n",
    "    path = f\"data/group/xjin/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.comm.tsv\"\n",
    "    try:\n",
    "        d = (\n",
    "            pd.read_table(path, index_col=[\"sample\", \"strain\"])\n",
    "            .squeeze()\n",
    "            .unstack()\n",
    "            # .rename(columns=str, index=lambda x: \"CF_\" + str(int(x.split(\"_\")[1])))\n",
    "            # .rename({'CF_11': 'CF_15', 'CF_15': 'CF_11'})  # Sample swap.\n",
    "        )\n",
    "    except FileNotFoundError:\n",
    "        missing_files.append(path)\n",
    "        d = pd.DataFrame([])\n",
    "    _keep_strains = idxwhere(d.sum() > 0.05)\n",
    "    assert d.index.isin(species_depth.index).all()\n",
    "    d = d.reindex(index=species_depth.index, columns=_keep_strains, fill_value=0)\n",
    "    d = d.assign(__other=lambda x: 1 - x.sum(1)).rename(columns={\"__other\": -1})\n",
    "    d[d < 0] = 0\n",
    "    d = d.divide(d.sum(1), axis=0)\n",
    "    d = d.multiply(species_depth[species_id], axis=0)\n",
    "    d = d.rename(columns=lambda s: f\"{species_id}_{s}\")\n",
    "    strain_depth.append(d)\n",
    "strain_depth = pd.concat(strain_depth, axis=1)\n",
    "strain_rabund = strain_depth.divide(strain_depth.sum(1), axis=0)\n",
    "len(species_depth.columns), len(missing_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xjin_strain_depth = strain_depth[lambda x: x.index.str.startswith(\"xjin_\")]\n",
    "xjin_strain_depth = xjin_strain_depth[idxwhere((xjin_strain_depth.sum() > 0.01))]\n",
    "plt.hist(np.log10(xjin_strain_depth.sum()), bins=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Number of benchmark samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xjin_strain_depth.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xjin_sotu_strain_list = idxwhere(xjin_strain_depth.sum() > 1)\n",
    "other_strain_depth = xjin_strain_depth.drop(columns=xjin_sotu_strain_list).sum(1)\n",
    "\n",
    "d = xjin_strain_depth[xjin_sotu_strain_list].assign(other=other_strain_depth)\n",
    "sns.clustermap(\n",
    "    d.T,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.1),\n",
    "    metric=\"cosine\",\n",
    "    figsize=(5, 8),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xjin_species_depth = species_depth[lambda x: x.index.str.startswith(\"xjin_\")].rename(\n",
    "    columns=str\n",
    ")\n",
    "xjin_species_depth = xjin_species_depth[idxwhere((xjin_species_depth.sum() > 0.01))]\n",
    "plt.hist(np.log10(xjin_species_depth.sum()), bins=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Depth Distribution of Species in Benchmark Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = xjin_species_depth.reindex(species_filt_list, axis=1, fill_value=0)\n",
    "print(d0.max().quantile([0.25, 0.5, 0.75]))\n",
    "_num_genomes_per_species = (\n",
    "    genome_to_spgc_strain.assign(species=genome.species_id)\n",
    "    .species.value_counts()\n",
    "    .reindex(\n",
    "        d0.columns, fill_value=0\n",
    "    )  # TODO: Do we only want to consider these 85 species, or the larger list?\n",
    ")\n",
    "_num_genomes_order = range(1, _num_genomes_per_species.max() + 1)\n",
    "_num_genomes_palette = lib.plot.construct_ordered_palette(\n",
    "    _num_genomes_order, cm=\"plasma\", other=\"white\"\n",
    ")\n",
    "_num_genomes_size_and_marker_palette = dict(\n",
    "    zip(_num_genomes_order, [(25, \"<\"), (20, \"s\"), (100, \"p\"), (150, \"*\")])\n",
    ")\n",
    "# _xjin_species_order = d.median().sort_values(ascending=False).index\n",
    "\n",
    "d1 = pd.DataFrame(\n",
    "    dict(\n",
    "        _max=d0.max(),\n",
    "        _sum=d0.sum(),\n",
    "        num_species_genomes=_num_genomes_per_species,\n",
    "        p__=d0.columns.map(species_taxonomy.p__),\n",
    "    )\n",
    ").assign(\n",
    "    depth_in_other_samples=lambda x: x._sum - x._max,\n",
    "    # num_genomes_c=lambda x: x.num_species_genomes.map(_num_genomes_palette),\n",
    "    phylum_c=lambda x: x.p__.map(phylum_palette),\n",
    "    # num_genomes_s=lambda x: 50 * np.sqrt(x.num_species_genomes)\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 4.5))\n",
    "\n",
    "for num_species_genomes in _num_genomes_order:\n",
    "    d2 = d1[lambda x: x.num_species_genomes == num_species_genomes]\n",
    "    markersize, markershape = _num_genomes_size_and_marker_palette[num_species_genomes]\n",
    "    ax.scatter(\n",
    "        \"_max\",\n",
    "        \"_sum\",\n",
    "        s=markersize,\n",
    "        facecolors=\"phylum_c\",\n",
    "        lw=0.5,\n",
    "        data=d2,\n",
    "        marker=markershape,\n",
    "        edgecolor=\"black\",\n",
    "        label=\"__nolegend__\",\n",
    "        alpha=0.8,\n",
    "    )\n",
    "    ax.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        s=markersize,\n",
    "        facecolors=\"grey\",\n",
    "        lw=0.5,\n",
    "        data=d2,\n",
    "        marker=markershape,\n",
    "        edgecolor=\"black\",\n",
    "        label=num_species_genomes,\n",
    "    )\n",
    "ax.plot([3e-2, 1e5], [3e-2, 1e5], lw=1, linestyle=\"--\", color=\"k\")\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xscale(\"log\")\n",
    "ax.set_xlim(3e-2, 1e3)\n",
    "ax.set_ylim(3e-2, 1e5)\n",
    "ax.set_aspect(1)\n",
    "ax.set_xlabel(\"Deepest Sample Depth\")\n",
    "ax.set_ylabel(\"Total Depth\")\n",
    "ax.legend(loc=\"lower right\", title=\"Num. Strains\")\n",
    "\n",
    "fig.savefig(\"fig/xjin_benchmark_species_details.svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d1.groupby('p__').num_species_genomes.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = xjin_species_depth.reindex(species_filt_list, axis=1, fill_value=0)\n",
    "print(d0.max().quantile([0.25, 0.5, 0.75]))\n",
    "_num_genomes_per_species = (\n",
    "    genome_to_spgc_strain.assign(species=genome.species_id)\n",
    "    .species.value_counts()\n",
    "    .reindex(\n",
    "        d0.columns, fill_value=0\n",
    "    )  # TODO: Do we only want to consider these 85 species, or the larger list?\n",
    ")\n",
    "_num_genomes_order = range(1, _num_genomes_per_species.max() + 1)\n",
    "_num_genomes_palette = lib.plot.construct_ordered_palette(\n",
    "    _num_genomes_order, cm=\"plasma\", other=\"white\"\n",
    ")\n",
    "_num_genomes_size_and_marker_palette = dict(\n",
    "    zip(_num_genomes_order, [(15, \"o\"), (30, \"s\"), (70, \"p\"), (120, \"*\")])\n",
    ")\n",
    "# _xjin_species_order = d.median().sort_values(ascending=False).index\n",
    "\n",
    "d1 = pd.DataFrame(\n",
    "    dict(\n",
    "        _max=d0.max(),\n",
    "        _sum=d0.sum(),\n",
    "        num_species_genomes=_num_genomes_per_species,\n",
    "        p__=d0.columns.map(species_taxonomy.p__),\n",
    "    )\n",
    ").assign(\n",
    "    depth_in_other_samples=lambda x: x._sum - x._max,\n",
    "    # num_genomes_c=lambda x: x.num_species_genomes.map(_num_genomes_palette),\n",
    "    phylum_c=lambda x: x.p__.map(phylum_palette),\n",
    "    # num_genomes_s=lambda x: 50 * np.sqrt(x.num_species_genomes)\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 4.5))\n",
    "\n",
    "for num_species_genomes in _num_genomes_order:\n",
    "    d2 = d1[lambda x: x.num_species_genomes == num_species_genomes]\n",
    "    markersize, markershape = _num_genomes_size_and_marker_palette[num_species_genomes]\n",
    "    ax.scatter(\n",
    "        \"_sum\",\n",
    "        \"_max\",\n",
    "        s=markersize,\n",
    "        facecolors=\"phylum_c\",\n",
    "        lw=0.5,\n",
    "        data=d2,\n",
    "        marker=markershape,\n",
    "        edgecolor=\"black\",\n",
    "        label=\"__nolegend__\",\n",
    "        alpha=0.85,\n",
    "    )\n",
    "    ax.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        s=markersize,\n",
    "        facecolors=\"grey\",\n",
    "        lw=0.5,\n",
    "        data=d2,\n",
    "        marker=markershape,\n",
    "        edgecolor=\"black\",\n",
    "        label=num_species_genomes,\n",
    "    )\n",
    "ax.plot([3e-2, 1e5], [3e-2, 1e5], lw=1, linestyle=\"--\", color=\"k\")\n",
    "ax.set_yscale(\"log\", subs=[])\n",
    "ax.set_xscale(\"log\", subs=[])\n",
    "ax.set_ylim(3e-2, 1e3)\n",
    "ax.set_xlim(3e-2, 1e5)\n",
    "ax.set_aspect(1)\n",
    "ax.set_xlabel(\"Total Depth\")\n",
    "ax.set_ylabel(\"Max Depth\")\n",
    "\n",
    "leg = ax.legend(loc=\"upper left\", title=\"Num. Strains\", labelspacing=0.1, frameon=False)\n",
    "leg._legend_box.align = \"left\"\n",
    "\n",
    "# fig.savefig(\"fig/xjin_benchmark_species_details.svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Number of strains per species after filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome.loc[genome_filt_list].species_id.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quality Control Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "tool = \"spgc-fit\"\n",
    "\n",
    "d = (\n",
    "    benchmark.xs((unit, match, tool), level=(\"unit\", \"match\", \"tool\"))\n",
    "    .reindex(genome_filt_list)\n",
    "    .join(genome_to_spgc_strain, rsuffix=\"_\")\n",
    "    .assign(log_species_depth_sum=lambda x: np.log10(x.species_depth_sum))[\n",
    "        lambda x: x.species_depth_sum > 0\n",
    "    ]\n",
    ")\n",
    "\n",
    "print(\n",
    "    d.shape\n",
    ")  # NOTE: There is one species (Collinsiella?) where strain deconvolution failed, and which therefore has no SPGC strains.\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    3, 3, squeeze=False, figsize=(10, 10), sharex=\"col\", sharey=\"row\"\n",
    ")\n",
    "\n",
    "for (x, y), ax in zip(\n",
    "    product(\n",
    "        [\n",
    "            \"log_selected_gene_depth_ratio_std\",\n",
    "            \"species_gene_frac\",\n",
    "            \"log_species_depth_sum\",\n",
    "        ],\n",
    "        [\"precision\", \"recall\", \"f1\"],\n",
    "    ),\n",
    "    axs.T.flatten(),\n",
    "):\n",
    "    print((x, y), sp.stats.spearmanr(d[x], d[y]))\n",
    "    ax.scatter(x, y, data=d, s=15, alpha=0.5)\n",
    "    ax.set_xlabel(x)\n",
    "    ax.set_ylabel(y)\n",
    "\n",
    "# (d1[lambda x: x[('precision', 'spgc-fit')] < 0.7])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "tool = \"spgc-fit\"\n",
    "\n",
    "d = (\n",
    "    benchmark.xs((unit, match, tool), level=(\"unit\", \"match\", \"tool\"))[\n",
    "        [\"precision\", \"recall\", \"f1\"]\n",
    "    ]\n",
    "    .reindex(genome_filt_list, fill_value=0)\n",
    "    .join(genome_to_spgc_strain, rsuffix=\"_\")\n",
    "    .assign(log_species_depth_sum=lambda x: np.log10(x.species_depth_sum))[\n",
    "        lambda x: x.species_depth_sum > 0\n",
    "    ]\n",
    ")\n",
    "low_quality_strains = idxwhere(~d.passes_filter)\n",
    "\n",
    "len(low_quality_strains), d.loc[low_quality_strains].f1.quantile(\n",
    "    [0.25, 0.5, 0.75]\n",
    "), d.drop(low_quality_strains).f1.quantile([0.25, 0.5, 0.75])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sp.stats.mannwhitneyu(d.loc[low_quality_strains].f1, d.drop(low_quality_strains).f1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1)\n",
    "\n",
    "plt.hist(d.loc[low_quality_strains].f1, bins=bins)\n",
    "plt.hist(d.drop(low_quality_strains).f1, bins=bins, alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.loc[low_quality_strains].precision.quantile([0.25, 0.5, 0.75]), d.drop(\n",
    "    low_quality_strains\n",
    ").precision.quantile([0.25, 0.5, 0.75])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.loc[low_quality_strains].recall.quantile([0.25, 0.5, 0.75]), d.drop(\n",
    "    low_quality_strains\n",
    ").recall.quantile([0.25, 0.5, 0.75])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sfacts",
   "language": "python",
   "name": "sfacts"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}