{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "_os.chdir(_os.environ['PROJECT_ROOT'])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import xarray as xr\n",
    "from lib.pandas_util import idxwhere, align_indexes, invert_mapping\n",
    "import matplotlib as mpl\n",
    "import lib.plot\n",
    "import statsmodels as sm\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "import subprocess\n",
    "from tempfile import mkstemp\n",
    "import time\n",
    "import subprocess\n",
    "from itertools import chain\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sfacts as sf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import lib.thisproject.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context('talk')\n",
    "plt.rcParams['figure.dpi'] = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_subset = 'xjin'\n",
    "group = 'xjin_hmp2'\n",
    "stemA = 'r.proc'\n",
    "\n",
    "path = {}\n",
    "\n",
    "path.update(dict(\n",
    "    species_taxonomy=\"ref/gtpro/species_taxonomy_ext.tsv\",\n",
    "    all_species_depth=f\"data/group/{group_subset}/{stemA}.gtpro.species_depth.tsv\",\n",
    "    midasdb_genomes=\"ref/uhgg_genomes_all_4644.tsv\",\n",
    "    strain_genomes=\"meta/genome.tsv\",\n",
    "))\n",
    "\n",
    "path_exists = {}\n",
    "for p in path:\n",
    "    path_exists[path[p]] = os.path.exists(path[p])\n",
    "\n",
    "assert all(path_exists.values()), '\\n'.join([\"Missing files:\"] + [p for p in path_exists if not path_exists[p]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_depth = lib.thisproject.data.load_species_depth(path['all_species_depth'])\n",
    "rabund = species_depth.apply(lambda x: x / x.sum(), axis=1)\n",
    "\n",
    "n_species = 90\n",
    "top_species = (rabund > 1e-5).sum().sort_values(ascending=False).head(n_species).index\n",
    "\n",
    "fig, axs = plt.subplots(n_species, figsize=(10, 0.5 * n_species), sharex=True, sharey=True)\n",
    "\n",
    "bins = np.logspace(-7, 0, num=51)\n",
    "\n",
    "for species_id, ax in zip(top_species, axs):\n",
    "    ax.hist(rabund[species_id], bins=bins, alpha=0.7)\n",
    "    ax.set_xscale('log')\n",
    "    prevalence = (rabund[species_id] > 1e-5).mean()\n",
    "    ax.set_title(\"\")\n",
    "    # ax.set_xticks()\n",
    "    # ax.set_yticks()\n",
    "    ax.yaxis.set_visible(False)\n",
    "    ax.xaxis.set_visible(False)\n",
    "    ax.patch.set_alpha(0.0)\n",
    "    for spine in ['left', 'right', 'top', 'bottom']:\n",
    "        ax.spines[spine].set_visible(False)\n",
    "    ax.annotate(f'{species_id} ({prevalence:0.0%})', xy=(0.05, 0.1), ha='left', xycoords=\"axes fraction\")\n",
    "    ax.set_xlim(left=1e-7)\n",
    "    ax.set_ylim(top=300)\n",
    "    \n",
    "ax.xaxis.set_visible(True)\n",
    "ax.spines['bottom'].set_visible(True)\n",
    "\n",
    "fig.subplots_adjust(hspace=-0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for species in sorted(top_species):\n",
    "    print(f'xjin_{species}_a', species, species_taxonomy.loc[species].g__, species_taxonomy.loc[species].s__, '', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = '101398'\n",
    "\n",
    "species_taxonomy = lib.thisproject.data.load_species_taxonomy(path[\"species_taxonomy\"])\n",
    "species_taxonomy.loc[species]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_genome = pd.read_table(path[\"strain_genomes\"], dtype='str')\n",
    "strain_genome[strain_genome.species_id == species]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_genome_ids = strain_genome[strain_genome.species_id == species].genome_id\n",
    "print(strain_genome_ids)\n",
    "strain_genome_id = strain_genome_ids.tolist()[0]\n",
    "assert strain_genome_ids.shape[0] == 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path.update(dict(\n",
    "    # uhgg_x_strain=f'data/species/sp-{species}/genome/midas_uhgg_pangenome.{strain_genome_id}-blastp.tsv',\n",
    "    strain_x_uhgg=f'data/species/sp-{species}/genome/{strain_genome_id}.midas_uhgg_pangenome-blastp.tsv',\n",
    "    strain_x_strain=f'data/species/sp-{species}/genome/{strain_genome_id}.{strain_genome_id}-blastp.tsv',\n",
    "))\n",
    "\n",
    "path_exists = {}\n",
    "for p in path:\n",
    "    path_exists[path[p]] = os.path.exists(path[p])\n",
    "\n",
    "assert all(path_exists.values()), '\\n'.join([\"Missing files:\"] + [p for p in path_exists if not path_exists[p]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Default file path forming for interactive use.\n",
    "\n",
    "stemB = 'filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts43-s85-seed0'\n",
    "stemC = 'sfacts42-seed0'\n",
    "spgc_params = 'e100'\n",
    "centroid = 75\n",
    "\n",
    "path.update(dict(\n",
    "    flag=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.{stemB}.refit-{stemC}.gene{centroid}.spgc-{spgc_params}.strain_files.flag\",\n",
    "    fit=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.{stemB}.world.nc\",\n",
    "    refit=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.{stemB}.refit-{stemC}.world.nc\",\n",
    "    strain_correlation=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.{stemB}.gene{centroid}.spgc-{spgc_params}.strain_correlation.tsv\",\n",
    "    strain_depth_ratio=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.{stemB}.gene{centroid}.spgc-{spgc_params}.strain_depth_ratio.tsv\",\n",
    "    strain_fraction=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.{stemB}.comm.tsv\",\n",
    "    species_gene_mean_depth=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.gene{centroid}.spgc-{spgc_params}.species_depth.tsv\",\n",
    "    species_gtpro_depth=f\"data/group/{group}/{stemA}.gtpro.species_depth.tsv\",\n",
    "    species_correlation=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.gene{centroid}.spgc-{spgc_params}.species_correlation.tsv\",\n",
    "    species_gene=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.gene{centroid}.spgc-{spgc_params}.species_gene.list\",\n",
    "    strain_thresholds=f\"data/group/{group}/species/sp-{species}/{stemA}.gtpro.{stemB}.gene{centroid}.spgc-{spgc_params}.strain_gene_threshold.tsv\",\n",
    "    gene_annotations=f\"ref/midasdb_uhgg_gene_annotations/sp-{species}.gene{centroid}_annotations.tsv\",\n",
    "    raw_gene_depth=f\"data/group/{group}/species/sp-{species}/{stemA}.gene{centroid}.depth.nc\",\n",
    "    reference_copy_number=f\"ref/midasdb_uhgg_pangenomes/{species}/gene{centroid}.reference_copy_number.nc\",\n",
    "    cluster_info=f\"ref/midasdb_uhgg/pangenomes/{species}/cluster_info.txt\",\n",
    "    gtpro_reference_genotype=f\"data/species/sp-{species}/gtpro_ref.mgtp.nc\",\n",
    "))\n",
    "\n",
    "path_exists = {}\n",
    "for p in path:\n",
    "    path_exists[path[p]] = os.path.exists(path[p])\n",
    "\n",
    "assert all(path_exists.values()), '\\n'.join([\"Missing files:\"] + [p for p in path_exists if not path_exists[p]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path['flag']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = sf.World.load(path['fit']).drop_low_abundance_strains(0.05)\n",
    "print(fit.sizes)\n",
    "position_ss = fit.random_sample(position=500).position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf.plot.plot_metagenotype(\n",
    "    fit.sel(position=position_ss),\n",
    "    # scaley=0.2, scalex=0.3,\n",
    "    row_linkage_func=lambda w: w.metagenotype.linkage(\"position\"),\n",
    "    col_linkage_func=lambda w: w.community.linkage(),\n",
    ")\n",
    "sf.plot.plot_depth(\n",
    "    fit.sel(position=position_ss),\n",
    "    # scaley=0.2, scalex=0.3,\n",
    "    row_linkage_func=lambda w: w.metagenotype.linkage(\"position\"),\n",
    "    col_linkage_func=lambda w: w.community.linkage(),\n",
    ")\n",
    "sf.plot.plot_dominance(\n",
    "    fit.sel(position=position_ss),\n",
    "    # scaley=0.2, scalex=0.3,\n",
    "    row_linkage_func=lambda w: w.metagenotype.linkage(\"position\"),\n",
    "    col_linkage_func=lambda w: w.community.linkage(),\n",
    ")\n",
    "sf.plot.plot_community(\n",
    "    fit.sel(position=position_ss),\n",
    "    # scaley=0.2, scalex=0.3,\n",
    "    col_linkage_func=lambda w: w.community.linkage(),\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(\"strain\"),\n",
    ")\n",
    "sf.plot.plot_genotype(\n",
    "    fit.sel(position=position_ss),\n",
    "    # scaley=0.2, scalex=0.3,\n",
    "    col_linkage_func=lambda w: w.metagenotype.linkage(\"position\"),\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(\"strain\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(fit.community.mean(\"sample\").to_series().sort_values(ascending=False).head(5))\n",
    "top_inferred_strain = fit.community.mean(\"sample\").to_series().idxmax()\n",
    "\n",
    "assert fit.community.mean(\"sample\").sel(strain=top_inferred_strain) > 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_cluster = pd.read_table(\n",
    "    path[\"cluster_info\"]\n",
    ").set_index('centroid_99', drop=False).rename_axis(index='gene_id')\n",
    "gene_annotation = pd.read_table(\n",
    "    path[\"gene_annotations\"],\n",
    "    names=['locus_tag', 'ftype', 'length_bp', 'gene', 'EC_number', 'COG', 'product'],\n",
    "    index_col='locus_tag',\n",
    ").rename(columns=str.lower)\n",
    "\n",
    "gene_meta = gene_cluster.loc[gene_cluster[f'centroid_{centroid}'].unique()].join(gene_annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blastp_header_names = [\n",
    "    'qseqid',\n",
    "    'sseqid',\n",
    "    'pident',\n",
    "    'length',\n",
    "    'mismatch',\n",
    "    'gapopen',\n",
    "    'qstart',\n",
    "    'qend',\n",
    "    'sstart',\n",
    "    'send',\n",
    "    'evalue',\n",
    "    'bitscore'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_strain_x_strain = (\n",
    "    pd.read_table(\n",
    "        path['strain_x_strain'],\n",
    "        names=blastp_header_names\n",
    "    )\n",
    ")\n",
    "\n",
    "_max_bitscore = _strain_x_strain.groupby(['qseqid']).bitscore.max()\n",
    "\n",
    "strain_x_uhgg = (\n",
    "    pd.read_table(\n",
    "        path['strain_x_uhgg'],\n",
    "        names=blastp_header_names\n",
    "    )\n",
    "    .assign(bitscore_ratio=lambda x: x.bitscore / x.qseqid.map(_max_bitscore))\n",
    "    .assign(sseq_centroid=lambda x: x.sseqid.map(gene_cluster[f'centroid_{centroid}']))\n",
    ")\n",
    "\n",
    "best_uhgg_hit = strain_x_uhgg.groupby('qseqid').apply(lambda d: d.sort_values('bitscore').iloc[-1]).groupby('sseq_centroid').bitscore_ratio.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(best_uhgg_hit, alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter('pident', 'bitscore_ratio', data=strain_x_uhgg, s=1)\n",
    "plt.plot([0, 100], [0, 1], color='k', linestyle='--')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "strain_thresholds = (\n",
    "    pd.read_table(path[\"strain_thresholds\"], index_col='strain')\n",
    "    .rename(columns=dict(\n",
    "        correlation_strict='corr_threshold_strict',\n",
    "        correlation_moderate='corr_threshold_moderate',\n",
    "        correlation_lenient='corr_threshold_lenient',\n",
    "        depth_high='depth_thresh_high',\n",
    "        depth_low='depth_thresh_low',\n",
    "    ))\n",
    ")\n",
    "\n",
    "strain_thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_corr = pd.read_table(path[\"strain_correlation\"], index_col=['gene_id', 'strain']).squeeze().unstack('strain', fill_value=0)\n",
    "strain_depth = pd.read_table(\n",
    "    path[\"strain_depth_ratio\"],\n",
    "    index_col=['gene_id', 'strain']\n",
    ").squeeze().unstack()\n",
    "# strain_corr, strain_depth = align_indexes(*align_indexes(strain_corr, strain_depth), axis=\"columns\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_thresholds = (\n",
    "    pd.read_table(path[\"strain_thresholds\"], index_col='strain')\n",
    "    .rename(columns=dict(\n",
    "        correlation_strict='corr_threshold_strict',\n",
    "        correlation_moderate='corr_threshold_moderate',\n",
    "        correlation_lenient='corr_threshold_lenient',\n",
    "        depth_high='depth_thresh_high',\n",
    "        depth_low='depth_thresh_low',\n",
    "    ))\n",
    ")\n",
    "\n",
    "_strain_meta = (\n",
    "    strain_thresholds\n",
    "    .join(fit.genotype.entropy().to_series().rename('genotype_entropy'))\n",
    "    # .join(refit.genotype.entropy().to_series().rename('genotype_refit_entropy'))\n",
    "    # .join(fit.metagenotype.entropy().to_series().rename('metagenotype_entropy').groupby(sample_to_strain).mean().rename(int))\n",
    "    # .join(strain_to_sample_list.apply(len).rename('num_samples'))\n",
    "    # .join(species_depth.apply(lambda x: x**(1)).groupby(sample_to_strain).std().rename('depth_stdev').rename(int))\n",
    "    # .join(species_depth.apply(lambda x: x**(1)).groupby(sample_to_strain).max().rename('depth_max').rename(int))\n",
    "    # .join(species_depth.apply(lambda x: x**(1)).groupby(sample_to_strain).sum().rename('depth_sum').rename(int))\n",
    "    # .assign(power_index=lambda x: (x.depth_stdev * np.sqrt(x.num_samples)).fillna(0))\n",
    ")\n",
    "strain_meta = _strain_meta\n",
    "\n",
    "\n",
    "# power_index_thresh = 5\n",
    "# genotype_entropy_thresh = 0.2\n",
    "# genotype_refit_entropy_thresh = 1.0\n",
    "\n",
    "# high_power_strain_list = idxwhere(\n",
    "#     (strain_meta.power_index > power_index_thresh)\n",
    "#     & (strain_meta.genotype_entropy < genotype_entropy_thresh)\n",
    "#     & (strain_meta.genotype_refit_entropy < genotype_refit_entropy_thresh)\n",
    "# )\n",
    "# print(len(high_power_strain_list))\n",
    "# highest_power_strain_list = strain_meta.sort_values('power_index', ascending=False).head(3).index\n",
    "\n",
    "# plt.scatter(strain_meta.power_index, strain_meta.corr_threshold_moderate, c=strain_meta.genotype_refit_entropy, alpha=0.5)\n",
    "# plt.axvline(power_index_thresh, lw=1, linestyle='--', color='k')\n",
    "# plt.colorbar()\n",
    "# plt.xscale('log')\n",
    "\n",
    "strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_corr = pd.read_table(path[\"species_correlation\"], names=['sample', 'correlation'], index_col='sample').squeeze()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "\n",
    "# depth_threshold = strain_meta.depth_thresh_low.loc[top_inferred_strain]\n",
    "# corr_threshold = strain_meta.corr_threshold_moderate.loc[top_inferred_strain]\n",
    "corr_threshold = 0.95  # Set manually, but this could/should be the automatically selected threshold.\n",
    "depth_threshold = 0.2  # Set manually, but this could/should be the automatically selected threshold.\n",
    "bitscore_threshold = 0.8\n",
    "\n",
    "strain_scores = (\n",
    "    pd.DataFrame(dict(\n",
    "        bitscore_ratio=best_uhgg_hit,\n",
    "        strain_corr=strain_corr[top_inferred_strain],\n",
    "        strain_depth=strain_depth[top_inferred_strain],\n",
    "        species_corr=species_corr,\n",
    "    ))\n",
    "    .fillna(0)\n",
    "    .assign(\n",
    "        bitscore_hit=lambda x: x.bitscore_ratio > bitscore_threshold,\n",
    "        corr_and_depth_hit=lambda x: (x.strain_corr > corr_threshold) & (x.strain_depth > depth_threshold),\n",
    "    )\n",
    "    .sort_values('bitscore_ratio')\n",
    ")\n",
    "\n",
    "\n",
    "# Color by bitscore_ratio\n",
    "fig, ax = plt.subplots()\n",
    "cax = make_axes_locatable(ax).append_axes('right', size='5%', pad=0.05)\n",
    "artist = ax.scatter(\n",
    "    'corr_complement',\n",
    "    'strain_depth',\n",
    "    data=strain_scores.assign(corr_complement=lambda x: 1 - x.strain_corr),\n",
    "    s=5,\n",
    "    c='bitscore_ratio',\n",
    "    # norm=mpl.colors.PowerNorm(1/5),\n",
    "    alpha=0.1\n",
    "    \n",
    ")\n",
    "ax.axhline(depth_threshold, lw=1, linestyle='--')\n",
    "ax.axhline(1, xmin=0., xmax=0.5, lw=1, linestyle='--', color='k')\n",
    "ax.axvline(1 - corr_threshold, lw=1, linestyle='--')\n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('symlog', linthresh=1e-2)\n",
    "ax.set_ylim(bottom=-1e-3)\n",
    "# ax.set_ylim(1e-2, 1e2)\n",
    "ax.invert_xaxis()\n",
    "\n",
    "cbar = fig.colorbar(artist, cax=cax)\n",
    "cbar.solids.set(alpha=1)\n",
    "\n",
    "# Color by species_corr\n",
    "fig, ax = plt.subplots()\n",
    "cax = make_axes_locatable(ax).append_axes('right', size='5%', pad=0.05)\n",
    "artist = ax.scatter(\n",
    "    'corr_complement',\n",
    "    'strain_depth',\n",
    "    data=strain_scores.assign(corr_complement=lambda x: 1 - x.strain_corr),\n",
    "    s=5,\n",
    "    c='corr_complement',\n",
    "    norm=mpl.colors.LogNorm(),\n",
    "    alpha=0.1,\n",
    "    cmap='viridis_r',\n",
    ")\n",
    "ax.axhline(depth_threshold, lw=1, linestyle='--')\n",
    "ax.axhline(1, xmin=0., xmax=0.5, lw=1, linestyle='--', color='k')\n",
    "ax.axvline(1 - corr_threshold, lw=1, linestyle='--')\n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('symlog', linthresh=1e-2)\n",
    "ax.set_ylim(bottom=-1e-3)\n",
    "# ax.set_ylim(1e-2, 1e2)\n",
    "ax.invert_xaxis()\n",
    "\n",
    "cbar = fig.colorbar(artist, cax=cax)\n",
    "cbar.solids.set(alpha=1)\n",
    "cax.invert_yaxis()\n",
    "\n",
    "\n",
    "print((strain_scores.bitscore_hit & strain_scores.corr_and_depth_hit).sum() / strain_scores.corr_and_depth_hit.sum())\n",
    "print((strain_scores.bitscore_hit & strain_scores.corr_and_depth_hit).sum() / strain_scores.bitscore_hit.sum())\n",
    "print(strain_scores.bitscore_hit.sum(), strain_scores.corr_and_depth_hit.sum(), (strain_scores.bitscore_hit & strain_scores.corr_and_depth_hit).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = strain_scores[lambda x: (x.bitscore_hit) & (x.strain_depth > 0)].join(gene_cluster.centroid_99_length).assign(\n",
    "        log_centroid_99_length=lambda x: np.log10(x.centroid_99_length / 3),\n",
    "        log_strain_depth=lambda x: np.log10(x.strain_depth),\n",
    "    )\n",
    "\n",
    "sns.regplot(\n",
    "    x='log_centroid_99_length',\n",
    "    y='log_strain_depth',\n",
    "    data=d,\n",
    "    lowess=True,\n",
    "    scatter_kws=dict(s=2),\n",
    ")\n",
    "plt.axhline(0, lw=1, linestyle='--', color='k')\n",
    "plt.ylim(-1.5, 1.5)\n",
    "# plt.xscale('log')\n",
    "# plt.yscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_annotation.loc[idxwhere(strain_scores.bitscore_hit)].join(strain_scores).loc[lambda x: x.strain_depth > 2][lambda x: x['product'] != 'hypothetical protein'].sort_values('strain_depth', ascending=False).head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_x_midas = strain_x_uhgg.groupby(['qseqid', 'sseq_centroid']).bitscore_ratio.max()\n",
    "\n",
    "\n",
    "bins = np.linspace(0, 1)\n",
    "plt.hist(orf_x_midas.unstack(fill_value=0).max(0), bins=bins, density=True)\n",
    "plt.hist(orf_x_midas.unstack(fill_value=0).max(1), bins=bins, density=True, alpha=0.5)\n",
    "plt.yscale('log')\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# _corr_threshold = 0.95  # Set manually, but this could/should be the automatically selected threshold.\n",
    "# _depth_threshold = 0.65  # Set manually, but this could/should be the automatically selected threshold.\n",
    "spgc_hit = idxwhere((strain_scores.strain_corr > corr_threshold) & (strain_scores.strain_depth > depth_threshold))\n",
    "\n",
    "spgc_hit_best_orf = orf_x_midas.unstack().reindex(columns=spgc_hit).fillna(0).max(0)\n",
    "orf_best_spgc_hit = orf_x_midas.unstack().reindex(columns=spgc_hit).fillna(0).max(1)\n",
    "orf_best_mapping = orf_x_midas.unstack(fill_value=0).max(1)\n",
    "\n",
    "bitscore_ratio_threshold = 0.8\n",
    "\n",
    "bins = np.linspace(0, 1)\n",
    "plt.hist(spgc_hit_best_orf, bins=bins, alpha=0.5, label='hits')\n",
    "plt.hist(orf_best_spgc_hit, bins=bins, alpha=0.5, label='orfs')\n",
    "plt.hist(orf_best_mapping, bins=bins, alpha=0.5, label='best')\n",
    "plt.yscale('log')\n",
    "plt.legend(loc='upper center')\n",
    "plt.axvline(bitscore_ratio_threshold, lw=1, linestyle='--', color='k')\n",
    "None\n",
    "\n",
    "# If we were to set a threshold bitscore (e.g. 0.8), then:\n",
    "# FP-rate would the fraction of the \"hits\" histogram left of that,\n",
    "# FN-rate would the fraction of the \"orfs\" histogram left of that,\n",
    "# A \"mapping failure\" rate would be the part of the \"best\" historgram\n",
    "# left of that. (Alternatively, I could simply remove these hits.)\n",
    "\n",
    "(\n",
    "    (spgc_hit_best_orf < bitscore_ratio_threshold).mean(),\n",
    "    (orf_best_spgc_hit < bitscore_ratio_threshold).mean(),\n",
    "    (orf_best_mapping < bitscore_ratio_threshold).mean(),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "midas_gene_list = idxwhere(strain_scores.strain_depth > 0)\n",
    "\n",
    "_correlation = strain_scores.strain_corr[midas_gene_list]\n",
    "_depth = strain_scores.strain_depth[midas_gene_list]\n",
    "_bitscore = orf_x_midas.unstack().reindex(columns=midas_gene_list, fill_value=0).astype(float)\n",
    "\n",
    "for c_thresh in [0, 0.25, 0.5, 0.75, 0.8, 0.9, 0.95, 0.99, 0.995]:\n",
    "    d_thresh = 0.25\n",
    "    b_thresh = 0.5\n",
    "    _bitscore_hit = _bitscore > b_thresh\n",
    "    _spgc_hit = (_correlation > c_thresh) & (_depth > d_thresh)\n",
    "\n",
    "    _match = (_spgc_hit & _bitscore_hit)\n",
    "\n",
    "    tp_midas = idxwhere(_bitscore_hit.loc[:, _spgc_hit].any(axis=0))  # MIDAS genes hit by both.\n",
    "    fp_midas = idxwhere(~(_bitscore_hit.loc[:, _spgc_hit].any()))  # MIDAS genes were hit by SPGC but never by BLAST?\n",
    "    fn_orf = idxwhere(~(_bitscore_hit.loc[:, _spgc_hit].any(axis=1)))  # How many ORFs were hit by BLAST but no matching SPGC hits?\n",
    "    tp_orf = idxwhere((_spgc_hit & _bitscore_hit).any(axis=1))  # How many ORFs were hit by BLAST and by SPGC?\n",
    "    \n",
    "    n_tp_midas = len(tp_midas)\n",
    "    n_fp_midas = len(fp_midas)\n",
    "    n_fn_orf = len(fn_orf)\n",
    "    n_tp_orf = len(tp_orf)\n",
    "\n",
    "    precision_midas = n_tp_midas / (n_tp_midas + n_fp_midas)\n",
    "    recall_orf = n_tp_orf / (n_tp_orf + n_fn_orf)\n",
    "\n",
    "    print(f'{c_thresh}\\t{d_thresh}\\t{precision_midas:0.2f}\\t{recall_orf:0.2f}\\t{n_fp_midas}\\t{n_fn_orf}\\t{n_tp_midas}\\t{n_tp_orf}')\n",
    "\n",
    "# MIDAS genes that map to multiple ORFs, tend to have higher depth ratios.\n",
    "#sns.regplot(x=_bitscore_hit.loc[tp_orf, tp_midas].sum(axis=0), y=strain_scores.loc[tp_midas].strain_depth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import product\n",
    "\n",
    "midas_gene_list = idxwhere((strain_scores.strain_corr > 0) | (strain_scores.strain_depth > 0))\n",
    "\n",
    "_correlation = strain_scores.strain_corr[midas_gene_list]\n",
    "_depth = strain_scores.strain_depth[midas_gene_list]\n",
    "_bitscore = orf_x_midas.unstack().reindex(columns=midas_gene_list, fill_value=0).astype(float)\n",
    "\n",
    "c_thresh_list_complement = np.logspace(-2, 0, num=10)\n",
    "c_thresh_list = list(reversed(1 - c_thresh_list_complement))\n",
    "d_thresh_list = np.linspace(0, 1.0, num=10)\n",
    "\n",
    "b_thresh = 0.8\n",
    "_bitscore_hit = _bitscore > b_thresh\n",
    "\n",
    "hits_matching = np.empty((len(d_thresh_list), len(c_thresh_list)))\n",
    "orfs_matching = np.empty_like(hits_matching)\n",
    "for (i, d_thresh), (j, c_thresh) in tqdm(product(enumerate(d_thresh_list), enumerate(c_thresh_list))):\n",
    "    _hits = (_correlation > c_thresh) & (_depth > d_thresh)\n",
    "    # hits_matching[i, j], orfs_matching[i, j] = two_way_accuracy(_bitscore_hit, _hits  # FIXME:\n",
    "    hits_matching[i, j] = (_bitscore_hit).any(axis=0)[_hits].mean()\n",
    "    orfs_matching[i, j] = (_bitscore_hit & _hits).any(axis=1).mean()\n",
    "    \n",
    "    \n",
    "fig, axs = plt.subplots(3, figsize=(5, 10))\n",
    "\n",
    "ax = axs[0]\n",
    "ax.set_title('hits_matching')\n",
    "cax = make_axes_locatable(ax).append_axes('right', size='5%', pad=0.05)\n",
    "artist = ax.pcolormesh(c_thresh_list_complement, d_thresh_list, hits_matching, norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1))\n",
    "ax.set_xscale('log')\n",
    "# ax = plt.gca()\n",
    "# ax.invert_xaxis()\n",
    "ax.set_xlim(right=1)\n",
    "# ax.set_ylim(0, 1)\n",
    "cbar = fig.colorbar(artist, cax=cax)\n",
    "\n",
    "ax = axs[1]\n",
    "ax.set_title('orfs_matching')\n",
    "cax = make_axes_locatable(ax).append_axes('right', size='5%', pad=0.05)\n",
    "artist = ax.pcolormesh(c_thresh_list_complement, d_thresh_list, orfs_matching, norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1))\n",
    "ax.set_xscale('log')\n",
    "ax.set_xlim(right=1)\n",
    "# ax.set_ylim(0, 1)\n",
    "cbar = fig.colorbar(artist, cax=cax)\n",
    "\n",
    "ax = axs[2]\n",
    "ax.set_title('product')\n",
    "cax = make_axes_locatable(ax).append_axes('right', size='5%', pad=0.05)\n",
    "artist = ax.pcolormesh(c_thresh_list_complement, d_thresh_list, hits_matching * orfs_matching, norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1))\n",
    "ax.set_xscale('log')\n",
    "ax.set_xlim(right=1)\n",
    "# ax.set_ylim(0, 1)\n",
    "cbar = fig.colorbar(artist, cax=cax)\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    orf_best_spgc_hit\n",
    "    [idxwhere(orf_best_spgc_hit < bitscore_ratio_threshold)]\n",
    "    .rename('bitscore_ratio')\n",
    "    .reset_index()\n",
    "    .assign(\n",
    "        contig=lambda x: x.qseqid.str.split('_').str[0].astype(int),\n",
    "        orf=lambda x: x.qseqid.str.split('_').str[1].astype(int),\n",
    "    )\n",
    "    .sort_values('orf')\n",
    ")\n",
    "\n",
    "d.head(50)\n",
    "\n",
    "# Many ORFs without a hit are contiguous."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = orf_x_midas.unstack().loc[idxwhere(orf_best_spgc_hit < bitscore_ratio_threshold)]\n",
    "d1 = d0[(d0.max(1) > 0)]\n",
    "missed_orfs_best_hit = d1.idxmax(1)\n",
    "\n",
    "d2 = gene_annotation.join(strain_scores).loc[missed_orfs_best_hit].assign(corr_complement=lambda x: 1 - x.strain_corr)\n",
    "\n",
    "\n",
    "# Color by bitscore_ratio\n",
    "fig, ax = plt.subplots()\n",
    "cax = make_axes_locatable(ax).append_axes('right', size='5%', pad=0.05)\n",
    "artist = ax.scatter(\n",
    "    'corr_complement',\n",
    "    'strain_depth',\n",
    "    data=d2,\n",
    "    s=10,\n",
    "    c='bitscore_ratio',\n",
    "    # norm=mpl.colors.PowerNorm(1/5),\n",
    "    alpha=0.9\n",
    "    \n",
    ")\n",
    "ax.axhline(depth_threshold, lw=1, linestyle='--')\n",
    "ax.axhline(1, xmin=0., xmax=0.5, lw=1, linestyle='--', color='k')\n",
    "ax.axvline(1 - corr_threshold, lw=1, linestyle='--')\n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('log')\n",
    "# ax.set_ylim(1e-2, 1e2)\n",
    "ax.invert_xaxis()\n",
    "\n",
    "fig.colorbar(artist, cax=cax)\n",
    "\n",
    "# TODO: Make a similar plot, but with the converse points (false positives)\n",
    "# showing whether we could have trimmed them off by adjusting our thresholds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sfacts",
   "language": "python",
   "name": "sfacts"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}